// ngx-scrollbar/fesm2022/ngx-scrollbar-addons.mjs
import * as i0 from '@angular/core';
import { inject, NgZone, input, numberAttribute, effect, Directive } from '@angular/core';
import { fromEvent, throttleTime, animationFrameScheduler, tap, debounceTime } from 'rxjs';
import { NG_SCROLLBAR_OPTIONS, ViewportAdapter } from 'ngx-scrollbar';

class NgScrollHoverVisibility {
    /** Global options */
    options = inject(NG_SCROLLBAR_OPTIONS);
    zone = inject(NgZone);
    viewport = inject(ViewportAdapter);
    /** Time in ms to wait after scrolling stops before hiding the scrollbar */
    scrollHideDelay = input(this.options.scrollHideDelay, { ...(ngDevMode ? { debugName: "scrollHideDelay" } : {}), transform: numberAttribute });
    /** Throttle time to prevent excessive updates during active scrolling */
    scrollThrottleTime = input(this.options.scrollThrottleTime, { ...(ngDevMode ? { debugName: "scrollThrottleTime" } : {}), transform: numberAttribute });
    constructor() {
        effect((onCleanup) => {
            let sub$;
            // Only activate this logic if the scrollbar is configured to show on hover
            if (this.viewport.initialized() && this.viewport.visibility() === 'hover') {
                const el = this.viewport.viewportElement;
                this.zone.runOutsideAngular(() => {
                    sub$ = fromEvent(el, 'scroll', { passive: true }).pipe(throttleTime(this.scrollThrottleTime(), animationFrameScheduler, {
                        leading: true,
                        trailing: false
                    }), tap(() => {
                        if (el.getAttribute('scrolling') !== 'true') {
                            el.setAttribute('scrolling', 'true');
                        }
                    }), debounceTime(this.scrollHideDelay()), tap(() => {
                        if (el.getAttribute('scrolling') !== 'false') {
                            el.setAttribute('scrolling', 'false');
                        }
                    })).subscribe();
                });
            }
            onCleanup(() => sub$?.unsubscribe());
        });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollHoverVisibility, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "21.1.2", type: NgScrollHoverVisibility, isStandalone: true, selector: "ng-scrollbar[visibleOnScroll]", inputs: { scrollHideDelay: { classPropertyName: "scrollHideDelay", publicName: "scrollHideDelay", isSignal: true, isRequired: false, transformFunction: null }, scrollThrottleTime: { classPropertyName: "scrollThrottleTime", publicName: "scrollThrottleTime", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollHoverVisibility, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-scrollbar[visibleOnScroll]'
                }]
        }], ctorParameters: () => [], propDecorators: { scrollHideDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "scrollHideDelay", required: false }] }], scrollThrottleTime: [{ type: i0.Input, args: [{ isSignal: true, alias: "scrollThrottleTime", required: false }] }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgScrollHoverVisibility };
