// ngx-scrollbar/fesm2022/ngx-scrollbar-reached-event.mjs
import * as i0 from '@angular/core';
import { inject, NgZone, output, input, viewChildren, afterNextRender, ChangeDetectionStrategy, Component, Renderer2, Injector, ApplicationRef, ElementRef, numberAttribute, EventEmitter, effect, createComponent, inputBinding, outputBinding, Output, Directive } from '@angular/core';
import { ViewportAdapter } from 'ngx-scrollbar';

class ReachDropObserver {
    zone = inject(NgZone);
    viewport = inject(ViewportAdapter);
    /** The intersection observer reference */
    intersectionObserver;
    events = output();
    selectedEvents = input(...(ngDevMode ? [undefined, { debugName: "selectedEvents" }] : []));
    triggerElements = viewChildren('detectElement', ...(ngDevMode ? [{ debugName: "triggerElements" }] : []));
    constructor() {
        afterNextRender({
            earlyRead: () => {
                this.zone.runOutsideAngular(() => {
                    // The first time the observer is triggered as soon as the element is observed,
                    // This flag is used to ignore this first trigger
                    let intersectionObserverInit = false;
                    this.intersectionObserver = new IntersectionObserver((entries) => {
                        if (intersectionObserverInit) {
                            entries.forEach((entry) => {
                                const entryType = entry.target.getAttribute('event');
                                if ((entryType === 'reached' && entry.isIntersecting) || (entryType === 'dropped' && !entry.isIntersecting)) {
                                    // Forward the detected trigger element only after the observer is initialized
                                    // Only observe the trigger elements when scrollable
                                    this.zone.run(() => this.events.emit(entry.target.getAttribute('name')));
                                }
                            });
                        }
                        else {
                            // Once the initial element is detected, set a flag to true
                            intersectionObserverInit = true;
                        }
                    }, {
                        root: this.viewport.viewportElement
                    });
                    this.triggerElements().forEach((el) => {
                        this.intersectionObserver.observe(el.nativeElement);
                    });
                });
            }
        });
    }
    ngOnDestroy() {
        this.intersectionObserver?.disconnect();
        this.intersectionObserver = null;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ReachDropObserver, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.1.2", type: ReachDropObserver, isStandalone: true, selector: "scroll-reached", inputs: { selectedEvents: { classPropertyName: "selectedEvents", publicName: "selectedEvents", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { events: "events" }, host: { properties: { "attr.isVerticallyScrollable": "viewport.isVerticallyScrollable()", "attr.isHorizontallyScrollable": "viewport.isHorizontallyScrollable()" } }, viewQueries: [{ propertyName: "triggerElements", predicate: ["detectElement"], descendants: true, isSignal: true }], ngImport: i0, template: `
    @for (event of selectedEvents(); track $index) {
      <div #detectElement
           class="scroll-trigger-element"
           [attr.name]="event.name"
           [attr.event]="event.type"
           [attr.trigger]="event.trigger"></div>
    }
  `, isInline: true, styles: [":host{--_reached-offset-top: var(--reached-offset-top, 1px);--_reached-offset-bottom: var(--reached-offset-bottom, 1px);--_reached-offset-start: var(--reached-offset-start, 1px);--_reached-offset-end: var(--reached-offset-end, 1px);--_dropped-offset-top: var(--dropped-offset-top, 1px);--_dropped-offset-bottom: var(--dropped-offset-bottom, 1px);--_dropped-offset-start: var(--dropped-offset-start, 1px);--_dropped-offset-end: var(--dropped-offset-end, 1px);display:block;position:absolute;inset:0;z-index:-9999;-webkit-user-select:none;user-select:none;pointer-events:none}:host[isHorizontallyScrollable=false] [trigger=end]{display:none}:host[isVerticallyScrollable=false] [trigger=bottom]{display:none}.scroll-trigger-element{position:absolute;-webkit-user-select:none;user-select:none;pointer-events:none;z-index:-9999;visibility:hidden}.scroll-trigger-element[event=reached]{background:red}.scroll-trigger-element[event=dropped]{background:#00f}.scroll-trigger-element:is([trigger=top],[trigger=bottom]){inset-inline-start:0;inset-inline-end:0}.scroll-trigger-element:is([trigger=start],[trigger=end]){inset-block-start:0;inset-block-end:0}.scroll-trigger-element[trigger=top]{inset-block-start:0}.scroll-trigger-element[trigger=bottom]{inset-block-end:0}.scroll-trigger-element[trigger=start]{inset-inline-start:0;inset-inline-end:unset}.scroll-trigger-element[trigger=end]{inset-inline-end:0;inset-inline-start:unset}.scroll-trigger-element[trigger=top][event=reached]{height:var(--_reached-offset-top)}.scroll-trigger-element[trigger=bottom][event=reached]{height:var(--_reached-offset-bottom)}.scroll-trigger-element[trigger=start][event=reached]{width:var(--_reached-offset-start)}.scroll-trigger-element[trigger=end][event=reached]{width:var(--_reached-offset-end)}.scroll-trigger-element[trigger=top][event=dropped]{height:var(--_dropped-offset-top)}.scroll-trigger-element[trigger=bottom][event=dropped]{height:var(--_dropped-offset-bottom)}.scroll-trigger-element[trigger=start][event=dropped]{width:var(--_dropped-offset-start)}.scroll-trigger-element[trigger=end][event=dropped]{width:var(--_dropped-offset-end)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ReachDropObserver, decorators: [{
            type: Component,
            args: [{ host: {
                        '[attr.isVerticallyScrollable]': 'viewport.isVerticallyScrollable()',
                        '[attr.isHorizontallyScrollable]': 'viewport.isHorizontallyScrollable()',
                    }, selector: 'scroll-reached', template: `
    @for (event of selectedEvents(); track $index) {
      <div #detectElement
           class="scroll-trigger-element"
           [attr.name]="event.name"
           [attr.event]="event.type"
           [attr.trigger]="event.trigger"></div>
    }
  `, changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{--_reached-offset-top: var(--reached-offset-top, 1px);--_reached-offset-bottom: var(--reached-offset-bottom, 1px);--_reached-offset-start: var(--reached-offset-start, 1px);--_reached-offset-end: var(--reached-offset-end, 1px);--_dropped-offset-top: var(--dropped-offset-top, 1px);--_dropped-offset-bottom: var(--dropped-offset-bottom, 1px);--_dropped-offset-start: var(--dropped-offset-start, 1px);--_dropped-offset-end: var(--dropped-offset-end, 1px);display:block;position:absolute;inset:0;z-index:-9999;-webkit-user-select:none;user-select:none;pointer-events:none}:host[isHorizontallyScrollable=false] [trigger=end]{display:none}:host[isVerticallyScrollable=false] [trigger=bottom]{display:none}.scroll-trigger-element{position:absolute;-webkit-user-select:none;user-select:none;pointer-events:none;z-index:-9999;visibility:hidden}.scroll-trigger-element[event=reached]{background:red}.scroll-trigger-element[event=dropped]{background:#00f}.scroll-trigger-element:is([trigger=top],[trigger=bottom]){inset-inline-start:0;inset-inline-end:0}.scroll-trigger-element:is([trigger=start],[trigger=end]){inset-block-start:0;inset-block-end:0}.scroll-trigger-element[trigger=top]{inset-block-start:0}.scroll-trigger-element[trigger=bottom]{inset-block-end:0}.scroll-trigger-element[trigger=start]{inset-inline-start:0;inset-inline-end:unset}.scroll-trigger-element[trigger=end]{inset-inline-end:0;inset-inline-start:unset}.scroll-trigger-element[trigger=top][event=reached]{height:var(--_reached-offset-top)}.scroll-trigger-element[trigger=bottom][event=reached]{height:var(--_reached-offset-bottom)}.scroll-trigger-element[trigger=start][event=reached]{width:var(--_reached-offset-start)}.scroll-trigger-element[trigger=end][event=reached]{width:var(--_reached-offset-end)}.scroll-trigger-element[trigger=top][event=dropped]{height:var(--_dropped-offset-top)}.scroll-trigger-element[trigger=bottom][event=dropped]{height:var(--_dropped-offset-bottom)}.scroll-trigger-element[trigger=start][event=dropped]{width:var(--_dropped-offset-start)}.scroll-trigger-element[trigger=end][event=dropped]{width:var(--_dropped-offset-end)}\n"] }]
        }], ctorParameters: () => [], propDecorators: { events: [{ type: i0.Output, args: ["events"] }], selectedEvents: [{ type: i0.Input, args: [{ isSignal: true, alias: "selectedEvents", required: false }] }], triggerElements: [{ type: i0.ViewChildren, args: ['detectElement', { isSignal: true }] }] } });

class NgScrollReachDrop {
    renderer = inject(Renderer2);
    injector = inject(Injector);
    appRef = inject(ApplicationRef);
    viewport = inject(ViewportAdapter);
    nativeElement = inject(ElementRef).nativeElement;
    /** Reached offset value in px */
    reachedTopOffset = input(0, { ...(ngDevMode ? { debugName: "reachedTopOffset" } : {}), transform: numberAttribute });
    reachedBottomOffset = input(0, { ...(ngDevMode ? { debugName: "reachedBottomOffset" } : {}), transform: numberAttribute });
    reachedStartOffset = input(0, { ...(ngDevMode ? { debugName: "reachedStartOffset" } : {}), transform: numberAttribute });
    reachedEndOffset = input(0, { ...(ngDevMode ? { debugName: "reachedEndOffset" } : {}), transform: numberAttribute });
    /** Dropped offset value in px */
    droppedTopOffset = input(0, { ...(ngDevMode ? { debugName: "droppedTopOffset" } : {}), transform: numberAttribute });
    droppedBottomOffset = input(0, { ...(ngDevMode ? { debugName: "droppedBottomOffset" } : {}), transform: numberAttribute });
    droppedStartOffset = input(0, { ...(ngDevMode ? { debugName: "droppedStartOffset" } : {}), transform: numberAttribute });
    droppedEndOffset = input(0, { ...(ngDevMode ? { debugName: "droppedEndOffset" } : {}), transform: numberAttribute });
    disableReached = input(false, ...(ngDevMode ? [{ debugName: "disableReached" }] : []));
    disableDropped = input(false, ...(ngDevMode ? [{ debugName: "disableDropped" }] : []));
    reachedTop = new EventEmitter();
    reachedBottom = new EventEmitter();
    reachedStart = new EventEmitter();
    reachedEnd = new EventEmitter();
    droppedTop = new EventEmitter();
    droppedBottom = new EventEmitter();
    droppedStart = new EventEmitter();
    droppedEnd = new EventEmitter();
    container;
    /** An array that contains the chosen outputs */
    events = [];
    constructor() {
        effect(() => {
            this.setCssVariable('--reached-offset-top', this.reachedTopOffset());
            this.setCssVariable('--reached-offset-bottom', this.reachedBottomOffset());
            this.setCssVariable('--reached-offset-start', this.reachedStartOffset());
            this.setCssVariable('--reached-offset-end', this.reachedEndOffset());
            this.setCssVariable('--dropped-offset-top', this.droppedTopOffset());
            this.setCssVariable('--dropped-offset-bottom', this.droppedBottomOffset());
            this.setCssVariable('--dropped-offset-start', this.droppedStartOffset());
            this.setCssVariable('--dropped-offset-end', this.droppedEndOffset());
        });
        effect((onCleanup) => {
            if (!this.viewport.initialized())
                return;
            if (!this.disableReached()) {
                if (this.reachedTop.observed) {
                    this.events.push({ name: 'reachedTop', trigger: 'top', type: 'reached' });
                }
                if (this.reachedBottom.observed) {
                    this.events.push({ name: 'reachedBottom', trigger: 'bottom', type: 'reached' });
                }
                if (this.reachedStart.observed) {
                    this.events.push({ name: 'reachedStart', trigger: 'start', type: 'reached' });
                }
                if (this.reachedEnd.observed) {
                    this.events.push({ name: 'reachedEnd', trigger: 'end', type: 'reached' });
                }
            }
            if (!this.disableDropped()) {
                if (this.droppedTop.observed) {
                    this.events.push({ name: 'droppedTop', trigger: 'top', type: 'dropped' });
                }
                if (this.droppedBottom.observed) {
                    this.events.push({ name: 'droppedBottom', trigger: 'bottom', type: 'dropped' });
                }
                if (this.droppedStart.observed) {
                    this.events.push({ name: 'droppedStart', trigger: 'start', type: 'dropped' });
                }
                if (this.droppedEnd.observed) {
                    this.events.push({ name: 'droppedEnd', trigger: 'end', type: 'dropped' });
                }
            }
            this.container = createComponent(ReachDropObserver, {
                elementInjector: this.injector,
                environmentInjector: this.appRef.injector,
                bindings: [
                    // Forward the observed outputs
                    inputBinding('selectedEvents', () => this.events),
                    // Forward the events to the appropriate output
                    outputBinding('events', (trigger) => this[trigger].emit()),
                ]
            });
            this.renderer.appendChild(this.viewport.contentWrapperElement, this.container.location.nativeElement);
            // Attach the host view of the component to the main change detection tree, so that its lifecycle hooks run.
            this.appRef.attachView(this.container.hostView);
            onCleanup(() => {
                this.events = [];
                this.container.destroy();
                this.container = null;
            });
        });
    }
    setCssVariable(property, value) {
        if (value) {
            this.nativeElement.style.setProperty(property, `${value}px`);
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollReachDrop, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "21.1.2", type: NgScrollReachDrop, isStandalone: true, selector: "\n    ng-scrollbar[reachedTop], ng-scrollbar[reachedBottom], ng-scrollbar[reachedStart], ng-scrollbar[reachedEnd],\n    ng-scrollbar[droppedTop], ng-scrollbar[droppedBottom], ng-scrollbar[droppedStart], ng-scrollbar[droppedEnd],\n    [ngScrollbar][reachedTop], [ngScrollbar][reachedBottom], [ngScrollbar][reachedStart], [ngScrollbar][reachedEnd],\n    [ngScrollbar][droppedTop], [ngScrollbar][droppedBottom], [ngScrollbar][droppedStart], [ngScrollbar][droppedEnd]\n  ", inputs: { reachedTopOffset: { classPropertyName: "reachedTopOffset", publicName: "reachedTopOffset", isSignal: true, isRequired: false, transformFunction: null }, reachedBottomOffset: { classPropertyName: "reachedBottomOffset", publicName: "reachedBottomOffset", isSignal: true, isRequired: false, transformFunction: null }, reachedStartOffset: { classPropertyName: "reachedStartOffset", publicName: "reachedStartOffset", isSignal: true, isRequired: false, transformFunction: null }, reachedEndOffset: { classPropertyName: "reachedEndOffset", publicName: "reachedEndOffset", isSignal: true, isRequired: false, transformFunction: null }, droppedTopOffset: { classPropertyName: "droppedTopOffset", publicName: "droppedTopOffset", isSignal: true, isRequired: false, transformFunction: null }, droppedBottomOffset: { classPropertyName: "droppedBottomOffset", publicName: "droppedBottomOffset", isSignal: true, isRequired: false, transformFunction: null }, droppedStartOffset: { classPropertyName: "droppedStartOffset", publicName: "droppedStartOffset", isSignal: true, isRequired: false, transformFunction: null }, droppedEndOffset: { classPropertyName: "droppedEndOffset", publicName: "droppedEndOffset", isSignal: true, isRequired: false, transformFunction: null }, disableReached: { classPropertyName: "disableReached", publicName: "disableReached", isSignal: true, isRequired: false, transformFunction: null }, disableDropped: { classPropertyName: "disableDropped", publicName: "disableDropped", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { reachedTop: "reachedTop", reachedBottom: "reachedBottom", reachedStart: "reachedStart", reachedEnd: "reachedEnd", droppedTop: "droppedTop", droppedBottom: "droppedBottom", droppedStart: "droppedStart", droppedEnd: "droppedEnd" }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollReachDrop, decorators: [{
            type: Directive,
            args: [{
                    selector: `
    ng-scrollbar[reachedTop], ng-scrollbar[reachedBottom], ng-scrollbar[reachedStart], ng-scrollbar[reachedEnd],
    ng-scrollbar[droppedTop], ng-scrollbar[droppedBottom], ng-scrollbar[droppedStart], ng-scrollbar[droppedEnd],
    [ngScrollbar][reachedTop], [ngScrollbar][reachedBottom], [ngScrollbar][reachedStart], [ngScrollbar][reachedEnd],
    [ngScrollbar][droppedTop], [ngScrollbar][droppedBottom], [ngScrollbar][droppedStart], [ngScrollbar][droppedEnd]
  `
                }]
        }], ctorParameters: () => [], propDecorators: { reachedTopOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "reachedTopOffset", required: false }] }], reachedBottomOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "reachedBottomOffset", required: false }] }], reachedStartOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "reachedStartOffset", required: false }] }], reachedEndOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "reachedEndOffset", required: false }] }], droppedTopOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "droppedTopOffset", required: false }] }], droppedBottomOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "droppedBottomOffset", required: false }] }], droppedStartOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "droppedStartOffset", required: false }] }], droppedEndOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "droppedEndOffset", required: false }] }], disableReached: [{ type: i0.Input, args: [{ isSignal: true, alias: "disableReached", required: false }] }], disableDropped: [{ type: i0.Input, args: [{ isSignal: true, alias: "disableDropped", required: false }] }], reachedTop: [{
                type: Output
            }], reachedBottom: [{
                type: Output
            }], reachedStart: [{
                type: Output
            }], reachedEnd: [{
                type: Output
            }], droppedTop: [{
                type: Output
            }], droppedBottom: [{
                type: Output
            }], droppedStart: [{
                type: Output
            }], droppedEnd: [{
                type: Output
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgScrollReachDrop };
