// ngx-scrollbar/fesm2022/ngx-scrollbar-mat.mjs
import * as i0 from '@angular/core';
import { inject, computed, Directive, effect } from '@angular/core';
import { MatSelect } from '@angular/material/select';
import { takeUntilDestroyed, outputToObservable } from '@angular/core/rxjs-interop';
import { ScrollbarInputOutputs, NgScrollbarAnywhere, ViewportAdapter } from 'ngx-scrollbar';
import { MatTimepicker } from '@angular/material/timepicker';
import { switchMap, of, fromEvent, tap } from 'rxjs';
import { MatAutocomplete } from '@angular/material/autocomplete';

class PopupScrollbar extends ScrollbarInputOutputs {
    scrollbarAnywhere = inject(NgScrollbarAnywhere);
    computedOptions = computed(() => {
        return {
            withButtons: this.withButtons(),
            position: this.position(),
            visibility: this.visibility(),
            appearance: this.appearance(),
            thumbClass: this.thumbClass(),
            trackClass: this.trackClass(),
            buttonClass: this.buttonClass(),
            hoverOffset: this.hoverOffset(),
            orientation: this.orientation(),
            disableSensor: this.disableSensor(),
            sensorThrottleTime: this.sensorThrottleTime(),
            disableInteraction: this.disableInteraction(),
            trackScrollDuration: this.trackScrollDuration()
        };
    }, ...(ngDevMode ? [{ debugName: "computedOptions" }] : []));
    scrollbarRef;
    afterInitSub;
    afterUpdateSub;
    init(id) {
        this.scrollbarRef = this.scrollbarAnywhere.createScrollbar(id, this.computedOptions());
        if (this.scrollbarRef) {
            const componentInstance = this.scrollbarRef.componentRef.instance;
            // The original CSS has an 8 px padding top and bottom on the viewport, this is already unset via CSS,
            // but we add the padding to the content wrapper
            componentInstance.nativeElement.style.setProperty('--_viewport-padding-block-start', '8px');
            componentInstance.nativeElement.style.setProperty('--_viewport-padding-block-end', '8px');
            this.afterInitSub = componentInstance.adapter.afterInit.subscribe(() => this.afterInit.emit());
            this.afterUpdateSub = componentInstance.adapter.afterUpdate.subscribe(() => this.afterUpdate.emit());
        }
    }
    destroy() {
        this.afterInitSub?.unsubscribe();
        this.afterUpdateSub?.unsubscribe();
        this.scrollbarRef?.destroy();
        this.scrollbarRef = null;
    }
}

class NgScrollbarMatSelect extends PopupScrollbar {
    matSelect = inject(MatSelect, { self: true });
    constructor() {
        super();
        this.matSelect.openedChange.pipe(takeUntilDestroyed()).subscribe((opened) => {
            if (opened) {
                this.init(`#${this.matSelect.id}-panel`);
            }
            else {
                this.destroy();
            }
        });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatSelect, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: NgScrollbarMatSelect, isStandalone: true, selector: "mat-select[matSelectScrollbar]", usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatSelect, decorators: [{
            type: Directive,
            args: [{
                    selector: 'mat-select[matSelectScrollbar]'
                }]
        }], ctorParameters: () => [] });

class NgScrollbarMatDialog {
    scrollbar = inject(ViewportAdapter);
    // NOTE: The only minor issue with this directive is that the scrollbar will be displayed in compact appearance
    constructor() {
        effect(() => {
            if (this.scrollbar.initialized()) {
                // Assign dialog-content padding by using its class on the content wrapper
                this.scrollbar.contentWrapperElement.classList.add('mat-mdc-dialog-content');
                // Remove harmful styles applied by the class
                this.scrollbar.contentWrapperElement.style.maxHeight = 'unset';
                this.scrollbar.contentWrapperElement.style.overflow = 'unset';
                this.scrollbar.contentWrapperElement.style.display = 'flow-root';
                // Check if dialog title exists
                if (this.scrollbar.viewportElement.parentElement.parentElement.querySelector('[mat-dialog-title], [matDialogTitle]')) {
                    this.scrollbar.contentWrapperElement.style.paddingTop = '0';
                }
            }
        });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatDialog, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: NgScrollbarMatDialog, isStandalone: true, selector: "mat-dialog-content[ngScrollbar]", host: { properties: { "style.padding": "0" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatDialog, decorators: [{
            type: Directive,
            args: [{
                    host: {
                        // Make sure the viewport doesn't have padding
                        '[style.padding]': '0'
                    },
                    selector: 'mat-dialog-content[ngScrollbar]',
                }]
        }], ctorParameters: () => [] });

class NgScrollbarMatTimepicker extends PopupScrollbar {
    matTimepicker = inject(MatTimepicker, { self: true });
    constructor() {
        super();
        outputToObservable(this.matTimepicker.opened).pipe(switchMap(() => {
            this.init(`#${this.matTimepicker.panelId}`);
            return outputToObservable(this.matTimepicker.closed);
        }), switchMap(() => {
            if (this.matTimepicker._animationsDisabled) {
                return of({});
            }
            // The MatTimepicker overlay has an exit animation by default, destroying the scrollbar before the timepicker; it won't get destroyed properly.
            return fromEvent(this.scrollbarRef.componentRef.instance.nativeElement, 'animationend');
        }), tap(() => this.destroy()), takeUntilDestroyed()).subscribe();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatTimepicker, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: NgScrollbarMatTimepicker, isStandalone: true, selector: "mat-timepicker[matTimepickerScrollbar]", usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatTimepicker, decorators: [{
            type: Directive,
            args: [{
                    selector: 'mat-timepicker[matTimepickerScrollbar]'
                }]
        }], ctorParameters: () => [] });

class NgScrollbarMatAutocomplete extends PopupScrollbar {
    matAutocomplete = inject(MatAutocomplete, { self: true });
    constructor() {
        super();
        this.matAutocomplete.opened.pipe(tap(() => this.init(`#${this.matAutocomplete.id}`)), switchMap(() => this.matAutocomplete.closed), tap(() => this.destroy()), takeUntilDestroyed()).subscribe();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatAutocomplete, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: NgScrollbarMatAutocomplete, isStandalone: true, selector: "mat-autocomplete[matAutocompleteScrollbar]", usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarMatAutocomplete, decorators: [{
            type: Directive,
            args: [{
                    selector: 'mat-autocomplete[matAutocompleteScrollbar]'
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgScrollbarMatAutocomplete, NgScrollbarMatDialog, NgScrollbarMatSelect, NgScrollbarMatTimepicker };
