// ngx-scrollbar/fesm2022/ngx-scrollbar.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, numberAttribute, output, Directive, signal, computed, ChangeDetectionStrategy, Component, DOCUMENT, afterNextRender, Injectable, NgZone, ElementRef, effect, untracked, afterRenderEffect, Renderer2, ApplicationRef, Injector, createComponent, Input, viewChild, linkedSignal, inputBinding, NgModule, makeEnvironmentProviders, EnvironmentInjector } from '@angular/core';
import { Directionality } from '@angular/cdk/bidi';
import { SmoothScrollManager } from 'ngx-scrollbar/smooth-scroll';
import { tap, throttleTime, fromEvent, startWith, switchMap, takeUntil, delay, filter, takeWhile, from, of, map, merge, interval, animationFrameScheduler, combineLatest } from 'rxjs';
import { Platform } from '@angular/cdk/platform';
import { SharedResizeObserver } from '@angular/cdk/observers/private';
import { ContentObserver } from '@angular/cdk/observers';
import { coerceElement } from '@angular/cdk/coercion';

let savedRanges = [];
function saveSelection(doc) {
    const selection = doc.getSelection();
    savedRanges = [];
    if (!selection?.rangeCount)
        return;
    for (let i = 0; i < selection.rangeCount; i++) {
        savedRanges.push(selection.getRangeAt(i).cloneRange());
    }
}
function restoreSelection(doc) {
    if (!savedRanges?.length)
        return;
    const selection = doc.getSelection();
    selection.removeAllRanges();
    savedRanges.forEach(range => selection.addRange(range));
    savedRanges = [];
}
function preventSelection(doc) {
    // Have to prevent default to avoid unexpected movement whe you grab object beneath scrollbar #476
    // https://github.com/MurhafSousli/ngx-scrollbar/issues/476
    return tap(() => {
        saveSelection(doc);
        doc.getSelection()?.removeAllRanges();
        doc.onselectstart = () => false;
    });
}
function enableSelection(doc) {
    return tap(() => {
        restoreSelection(doc);
        doc.onselectstart = null;
    });
}
function stopPropagation() {
    return tap((e) => e.stopPropagation());
}
function getThrottledStream(stream, duration) {
    return !duration ? stream : stream.pipe(throttleTime(duration, null, {
        leading: false,
        trailing: true
    }));
}
var ViewportClasses;
(function (ViewportClasses) {
    ViewportClasses["Viewport"] = "ng-scroll-viewport";
    ViewportClasses["Content"] = "ng-scroll-content";
    ViewportClasses["Spacer"] = "ng-scroll-spacer";
})(ViewportClasses || (ViewportClasses = {}));

const defaultOptions = {
    trackClass: '',
    thumbClass: '',
    buttonClass: '',
    orientation: 'auto',
    appearance: 'native',
    visibility: 'native',
    position: 'native',
    trackScrollDuration: 50,
    sensorThrottleTime: 0,
    disableSensor: false,
    disableInteraction: false,
    withButtons: false,
    hoverOffset: false,
    scrollHideDelay: 400,
    scrollThrottleTime: 200
};
// This CDN link for a modified version of the polyfill to fix firefox bug https://github.com/MurhafSousli/ngx-scrollbar/issues/615
const defaultScrollTimelinePolyfill = 'https://cdn.jsdelivr.net/gh/MurhafSousli/ngx-scrollbar@19.1.0/projects/ngx-scrollbar/src/assets/scroll-timeline-polyfill.js';

var ScrollbarUpdateReason;
(function (ScrollbarUpdateReason) {
    ScrollbarUpdateReason["AfterInit"] = "AfterInit";
    ScrollbarUpdateReason["Resized"] = "ResizeObserver";
})(ScrollbarUpdateReason || (ScrollbarUpdateReason = {}));
const NG_SCROLLBAR_OPTIONS = new InjectionToken('NG_SCROLLBAR_OPTIONS', {
    providedIn: 'root',
    factory: () => defaultOptions
});
const NG_SCROLLBAR_POLYFILL = new InjectionToken('NG_SCROLLBAR_POLYFILL', {
    providedIn: 'root',
    factory: () => defaultScrollTimelinePolyfill
});

class ScrollbarInputOutputs {
    /** Global options */
    options = inject(NG_SCROLLBAR_OPTIONS);
    /**
     * Sets the supported scroll track of the viewport, there are 3 options:
     *
     * - `vertical` Use both vertical and horizontal scrollbar
     * - `horizontal` Use both vertical and horizontal scrollbar
     * - `auto` Use both vertical and horizontal scrollbar
     */
    orientation = input(this.options.orientation, ...(ngDevMode ? [{ debugName: "orientation" }] : []));
    /**
     * When to show the scrollbar, and there are 3 options:
     *
     * - `native` (default) Scrollbar will be visible when viewport is scrollable like with native scrollbar
     * - `hover` Scrollbars are hidden by default, only visible on scrolling or hovering
     * - `visible` Scrollbars are always visible even if the viewport is not scrollable
     */
    visibility = input(this.options.visibility, ...(ngDevMode ? [{ debugName: "visibility" }] : []));
    /** Show scrollbar buttons */
    withButtons = input(this.options.withButtons, { ...(ngDevMode ? { debugName: "withButtons" } : {}), transform: booleanAttribute });
    /** Disables scrollbar interaction like dragging thumb and jumping by track click */
    disableInteraction = input(this.options.disableInteraction, { ...(ngDevMode ? { debugName: "disableInteraction" } : {}), transform: booleanAttribute });
    /** Whether ResizeObserver is disabled */
    disableSensor = input(this.options.disableSensor, { ...(ngDevMode ? { debugName: "disableSensor" } : {}), transform: booleanAttribute });
    /** Throttle interval for detecting changes via ResizeObserver */
    sensorThrottleTime = input(this.options.sensorThrottleTime, { ...(ngDevMode ? { debugName: "sensorThrottleTime" } : {}), transform: numberAttribute });
    /** A flag used to activate hover effect on the offset area around the scrollbar */
    hoverOffset = input(this.options.hoverOffset, { ...(ngDevMode ? { debugName: "hoverOffset" } : {}), transform: booleanAttribute });
    /** Scroll duration when the scroll track is clicked */
    trackScrollDuration = input(this.options.trackScrollDuration, { ...(ngDevMode ? { debugName: "trackScrollDuration" } : {}), transform: numberAttribute });
    /**
     *  Sets the appearance of the scrollbar, there are 2 options:
     *
     * - `native` (default) scrollbar space will be reserved just like with native scrollbar.
     * - `compact` scrollbar doesn't reserve any space, they are placed over the viewport.
     */
    appearance = input(this.options.appearance, ...(ngDevMode ? [{ debugName: "appearance" }] : []));
    /**
     * Sets the position of each scrollbar, there are 4 options:
     *
     * - `native` (Default) Use the default position like in native scrollbar.
     * - `invertY` Inverts vertical scrollbar position
     * - `invertX` Inverts Horizontal scrollbar position
     * - `invertAll` Inverts both scrollbar positions
     */
    position = input(this.options.position, ...(ngDevMode ? [{ debugName: "position" }] : []));
    /** A class forwarded to the scrollbar track element */
    trackClass = input(this.options.trackClass, ...(ngDevMode ? [{ debugName: "trackClass" }] : []));
    /** A class forwarded to the scrollbar thumb element */
    thumbClass = input(this.options.thumbClass, ...(ngDevMode ? [{ debugName: "thumbClass" }] : []));
    /** A class forwarded to the scrollbar button element */
    buttonClass = input(this.options.thumbClass, ...(ngDevMode ? [{ debugName: "buttonClass" }] : []));
    /** Steam that emits when scrollbar is initialized */
    afterInit = output();
    /** Steam that emits when scrollbar is updated */
    afterUpdate = output();
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarInputOutputs, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "21.1.2", type: ScrollbarInputOutputs, isStandalone: true, inputs: { orientation: { classPropertyName: "orientation", publicName: "orientation", isSignal: true, isRequired: false, transformFunction: null }, visibility: { classPropertyName: "visibility", publicName: "visibility", isSignal: true, isRequired: false, transformFunction: null }, withButtons: { classPropertyName: "withButtons", publicName: "withButtons", isSignal: true, isRequired: false, transformFunction: null }, disableInteraction: { classPropertyName: "disableInteraction", publicName: "disableInteraction", isSignal: true, isRequired: false, transformFunction: null }, disableSensor: { classPropertyName: "disableSensor", publicName: "disableSensor", isSignal: true, isRequired: false, transformFunction: null }, sensorThrottleTime: { classPropertyName: "sensorThrottleTime", publicName: "sensorThrottleTime", isSignal: true, isRequired: false, transformFunction: null }, hoverOffset: { classPropertyName: "hoverOffset", publicName: "hoverOffset", isSignal: true, isRequired: false, transformFunction: null }, trackScrollDuration: { classPropertyName: "trackScrollDuration", publicName: "trackScrollDuration", isSignal: true, isRequired: false, transformFunction: null }, appearance: { classPropertyName: "appearance", publicName: "appearance", isSignal: true, isRequired: false, transformFunction: null }, position: { classPropertyName: "position", publicName: "position", isSignal: true, isRequired: false, transformFunction: null }, trackClass: { classPropertyName: "trackClass", publicName: "trackClass", isSignal: true, isRequired: false, transformFunction: null }, thumbClass: { classPropertyName: "thumbClass", publicName: "thumbClass", isSignal: true, isRequired: false, transformFunction: null }, buttonClass: { classPropertyName: "buttonClass", publicName: "buttonClass", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { afterInit: "afterInit", afterUpdate: "afterUpdate" }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarInputOutputs, decorators: [{
            type: Directive
        }], propDecorators: { orientation: [{ type: i0.Input, args: [{ isSignal: true, alias: "orientation", required: false }] }], visibility: [{ type: i0.Input, args: [{ isSignal: true, alias: "visibility", required: false }] }], withButtons: [{ type: i0.Input, args: [{ isSignal: true, alias: "withButtons", required: false }] }], disableInteraction: [{ type: i0.Input, args: [{ isSignal: true, alias: "disableInteraction", required: false }] }], disableSensor: [{ type: i0.Input, args: [{ isSignal: true, alias: "disableSensor", required: false }] }], sensorThrottleTime: [{ type: i0.Input, args: [{ isSignal: true, alias: "sensorThrottleTime", required: false }] }], hoverOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "hoverOffset", required: false }] }], trackScrollDuration: [{ type: i0.Input, args: [{ isSignal: true, alias: "trackScrollDuration", required: false }] }], appearance: [{ type: i0.Input, args: [{ isSignal: true, alias: "appearance", required: false }] }], position: [{ type: i0.Input, args: [{ isSignal: true, alias: "position", required: false }] }], trackClass: [{ type: i0.Input, args: [{ isSignal: true, alias: "trackClass", required: false }] }], thumbClass: [{ type: i0.Input, args: [{ isSignal: true, alias: "thumbClass", required: false }] }], buttonClass: [{ type: i0.Input, args: [{ isSignal: true, alias: "buttonClass", required: false }] }], afterInit: [{ type: i0.Output, args: ["afterInit"] }], afterUpdate: [{ type: i0.Output, args: ["afterUpdate"] }] } });

const INITIAL_DIMENSION_VALUE = { width: 0, height: 0 };
/**
 * Class representing a viewport adapter.
 * Provides methods and properties to interact with a viewport and its content.
 */
class ViewportAdapter extends ScrollbarInputOutputs {
    smoothScroll = inject(SmoothScrollManager);
    /**
     * Indicates if the direction is 'ltr' or 'rtl'
     */
    dir = inject(Directionality);
    /**
     * Indicates when scrollbar thumb is being dragged
     */
    dragging = signal('none', ...(ngDevMode ? [{ debugName: "dragging" }] : []));
    state = computed(() => {
        let verticalUsed = false;
        let horizontalUsed = false;
        let isVerticallyScrollable = false;
        let isHorizontallyScrollable = false;
        const orientation = this.orientation();
        const visibility = this.visibility();
        const viewportDimensions = this.viewportDimension();
        const contentDimensions = this.contentDimension();
        // Check if the vertical scrollbar should be displayed
        if (orientation === 'auto' || orientation === 'vertical') {
            isVerticallyScrollable = contentDimensions.height > viewportDimensions.height;
            verticalUsed = visibility === 'visible' || isVerticallyScrollable;
        }
        // Check if the horizontal scrollbar should be displayed
        if (orientation === 'auto' || orientation === 'horizontal') {
            isHorizontallyScrollable = contentDimensions.width > viewportDimensions.width;
            horizontalUsed = visibility === 'visible' || isHorizontallyScrollable;
        }
        return {
            verticalUsed,
            horizontalUsed,
            isVerticallyScrollable,
            isHorizontallyScrollable,
        };
    }, ...(ngDevMode ? [{ debugName: "state" }] : []));
    isVerticallyScrollable = computed(() => this.state().isVerticallyScrollable, ...(ngDevMode ? [{ debugName: "isVerticallyScrollable" }] : []));
    isHorizontallyScrollable = computed(() => this.state().isHorizontallyScrollable, ...(ngDevMode ? [{ debugName: "isHorizontallyScrollable" }] : []));
    verticalUsed = computed(() => this.state().verticalUsed, ...(ngDevMode ? [{ debugName: "verticalUsed" }] : []));
    horizontalUsed = computed(() => this.state().horizontalUsed, ...(ngDevMode ? [{ debugName: "horizontalUsed" }] : []));
    /** Viewport dimension */
    viewportDimension = signal(INITIAL_DIMENSION_VALUE, ...(ngDevMode ? [{ debugName: "viewportDimension" }] : []));
    /** Content dimension */
    contentDimension = signal(INITIAL_DIMENSION_VALUE, ...(ngDevMode ? [{ debugName: "contentDimension" }] : []));
    /**
     * Viewport native element
     */
    viewportElement;
    /**
     * The element that wraps the content inside the viewport,
     * used to measure the content size and observe its changes.
     */
    contentWrapperElement;
    /*
     * A signal that indicates when viewport adapter is initialized
     */
    initialized = signal(false, ...(ngDevMode ? [{ debugName: "initialized" }] : []));
    /** Viewport clientHeight */
    get clientHeight() {
        return this.viewportElement.clientHeight;
    }
    /** Viewport clientWidth */
    get clientWidth() {
        return this.viewportElement.clientWidth;
    }
    /** Viewport scrollTop */
    get scrollTop() {
        return this.viewportElement.scrollTop;
    }
    /** Viewport scrollLeft */
    get scrollLeft() {
        return this.viewportElement.scrollLeft;
    }
    /** Content height */
    get contentHeight() {
        return this.contentWrapperElement.clientHeight;
    }
    /** Content width */
    get contentWidth() {
        return this.contentWrapperElement.clientWidth;
    }
    /** The remaining vertical scrollable distance. */
    get scrollMaxX() {
        return this.contentWidth - this.clientWidth;
    }
    /** The vertical remaining scrollable distance */
    get scrollMaxY() {
        return this.contentHeight - this.clientHeight;
    }
    /**
     * Initialize viewport
     */
    init(viewportElement, contentElement, spacerElement) {
        this.viewportElement = viewportElement;
        // When integrating the scrollbar with virtual scroll, the content wrapper will have fake size,
        // and a spacer element will have the real size
        // Therefore, if spaceElement is provided, it will be observed instead of the content wrapper
        if (spacerElement) {
            spacerElement.classList.add(ViewportClasses.Spacer);
            this.contentWrapperElement = spacerElement;
        }
        else {
            this.contentWrapperElement = contentElement;
        }
        this.initialized.set(true);
    }
    reset() {
        this.viewportElement = null;
        this.contentWrapperElement = null;
        this.initialized.set(false);
    }
    /**
     * Scrolls the viewport vertically to the specified value.
     */
    scrollYTo(value) {
        this.viewportElement.scrollTop = value;
    }
    /**
     * Scrolls the viewport horizontally to the specified value.
     */
    scrollXTo(value) {
        this.viewportElement.scrollLeft = value;
    }
    /**
     * Smooth scroll functions
     */
    scrollTo(options) {
        return this.smoothScroll.scrollTo(this.viewportElement, options);
    }
    /**
     * Scroll to an element by reference or selector
     */
    scrollToElement(target, options) {
        return this.smoothScroll.scrollToElement(this.viewportElement, target, options);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ViewportAdapter, deps: null, target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: ViewportAdapter, isStandalone: true, usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ViewportAdapter, decorators: [{
            type: Directive
        }] });

class ScrollContent {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollContent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.1.2", type: ScrollContent, isStandalone: true, selector: "ng-scroll-content", host: { properties: { "class.ng-scroll-content": "true" } }, ngImport: i0, template: '<ng-content/>', isInline: true, styles: [":host{box-sizing:border-box;width:var(--_scrollbar-content-width);z-index:1;min-width:100%;min-height:100%;padding-block-start:var(--_viewport-padding-block-start, 0);padding-block-end:var(--_viewport-padding-block-end, 0);padding-inline-start:var(--_viewport-padding-inline-start, 0);padding-inline-end:var(--_viewport-padding-inline-end, 0)}@layer default{:host{position:relative}:host:not(table){display:flow-root}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollContent, decorators: [{
            type: Component,
            args: [{ host: {
                        '[class.ng-scroll-content]': 'true'
                    }, selector: 'ng-scroll-content', template: '<ng-content/>', changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{box-sizing:border-box;width:var(--_scrollbar-content-width);z-index:1;min-width:100%;min-height:100%;padding-block-start:var(--_viewport-padding-block-start, 0);padding-block-end:var(--_viewport-padding-block-end, 0);padding-inline-start:var(--_viewport-padding-inline-start, 0);padding-inline-end:var(--_viewport-padding-inline-end, 0)}@layer default{:host{position:relative}:host:not(table){display:flow-root}}\n"] }]
        }] });

const SCROLLBAR_CONTROL = new InjectionToken('SCROLLBAR_CONTROL');
class ScrollbarAdapter {
    // Host component reference
    cmp = inject(ViewportAdapter);
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarAdapter, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: ScrollbarAdapter, isStandalone: true, host: { properties: { "class.ng-scrollbar-sticky": "true", "class.ng-scrollbar-hover": "cmp.hoverOffset()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarAdapter, decorators: [{
            type: Directive,
            args: [{
                    host: {
                        '[class.ng-scrollbar-sticky]': 'true',
                        '[class.ng-scrollbar-hover]': 'cmp.hoverOffset()'
                    }
                }]
        }] });

class ScrollbarManager {
    _polyfillUrl = inject(NG_SCROLLBAR_POLYFILL);
    document = inject(DOCUMENT);
    window = this.document.defaultView;
    scrollTimelinePolyfill = signal(null, ...(ngDevMode ? [{ debugName: "scrollTimelinePolyfill" }] : []));
    constructor() {
        afterNextRender({
            earlyRead: () => {
                if (!this.window['ScrollTimeline'] || !CSS.supports('animation-timeline', 'scroll()')) {
                    this.initPolyfill();
                }
            }
        });
    }
    async initPolyfill() {
        try {
            // Create a script element
            const script = this.document.createElement('script');
            script.src = this._polyfillUrl;
            script.async = true;
            // Wait for the script to load
            await new Promise((resolve, reject) => {
                script.onload = resolve;
                script.onerror = reject;
                this.document.head.appendChild(script);
            });
            // Once loaded, access and execute the function attached to the window object
            if (this.window['ScrollTimeline']) {
                this.scrollTimelinePolyfill.set(this.window['ScrollTimeline']);
            }
            else {
                console.error('[NgScrollbar]: Polyfill script loaded but ScrollTimeline not found.');
            }
        }
        catch (error) {
            console.error('[NgScrollbar]: Error loading ScrollTimeline script:', error);
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarManager, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarManager, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarManager, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: () => [] });

class PointerEventsAdapter {
    // Reference to the ScrollViewport component
    adapter = inject(ViewportAdapter);
    // Reference to the Scrollbar control component
    control = inject(SCROLLBAR_CONTROL);
    // Reference to the Document element
    document = inject(DOCUMENT);
    // Reference to angular zone
    zone = inject(NgZone);
    // The native element of the directive
    nativeElement = inject((ElementRef)).nativeElement;
    // Pointer events subscription (made public for testing purpose)
    _pointerEventsSub;
    constructor() {
        effect((onCleanup) => {
            const disableInteraction = this.adapter.disableInteraction();
            untracked(() => {
                if (!disableInteraction) {
                    this.zone.runOutsideAngular(() => {
                        this._pointerEventsSub = this.pointerEvents.subscribe();
                    });
                }
                onCleanup(() => this._pointerEventsSub?.unsubscribe());
            });
        });
    }
}

class TrackAdapter extends PointerEventsAdapter {
    pointerOut$ = fromEvent(this.nativeElement, 'pointerout', { passive: true });
    // The current position of the mouse during track dragging
    currMousePosition;
    // The direction of scroll when the track area is clicked
    scrollDirection;
    // The maximum scroll position until the end is reached
    scrollMax;
    // Returns viewport client size
    get viewportSize() {
        return this.adapter[this.control.sizeProperty];
    }
    // Get track client rect
    get clientRect() {
        return this.nativeElement.getBoundingClientRect();
    }
    // Scrollbar track offset
    get offset() {
        return this.clientRect[this.control.rectOffsetProperty];
    }
    // Scrollbar track length
    get size() {
        // Noticed that clientHeight is evaluated before getClientRect.height,
        // causing a wrong track size when integrated in dropdown integration
        return this.nativeElement[this.control.sizeProperty];
    }
    // Observable for track dragging events
    get pointerEvents() {
        // Observable streams for pointer events
        const pointerDown$ = fromEvent(this.nativeElement, 'pointerdown').pipe(stopPropagation(), preventSelection(this.document));
        const pointerUp$ = fromEvent(this.document, 'pointerup', { passive: true }).pipe(enableSelection(this.document));
        const pointerOver$ = fromEvent(this.nativeElement, 'pointerover', { passive: true }).pipe(
        // When the mouse is out and enters again, must set the current position first
        tap((e) => this.currMousePosition = this.getPointerPosition(e)), startWith({}));
        // Keep track of the current mouse location while dragging
        const pointerMove$ = fromEvent(this.nativeElement, 'pointermove', { passive: true }).pipe(tap((e) => this.currMousePosition = this.getPointerPosition(e)));
        return pointerDown$.pipe(switchMap((startEvent) => {
            // Track pointer location while dragging
            pointerMove$.pipe(takeUntil(pointerUp$)).subscribe();
            return this.onTrackFirstClick(startEvent).pipe(delay(200), switchMap(() => {
                // Otherwise, activate pointermove and pointerout events and switch to ongoing scroll calls
                return pointerOver$.pipe(filter(() => this.scrollDirection === this.getScrollDirection(this.currMousePosition)), switchMap(() => this.onTrackOngoingMousedown()));
            }), takeUntil(pointerUp$));
        }));
    }
    constructor() {
        super();
    }
    getPointerPosition(e) {
        return e[this.control.clientProperty] - this.offset;
    }
    /**
     *  Callback when mouse is first clicked on the track
     */
    onTrackFirstClick(e) {
        // Initialize variables and determine scroll direction
        this.currMousePosition = this.getPointerPosition(e);
        this.scrollDirection = this.getScrollDirection(this.currMousePosition);
        this.scrollMax = this.control.viewportScrollMax;
        return this.scrollTo(this.nextStep());
    }
    nextStep() {
        // Check which direction should the scroll go (forward or backward)
        if (this.scrollDirection === 'forward') {
            // Scroll forward
            const scrollForwardIncrement = this.getScrollForwardStep();
            // Check if the incremental position is bigger than the scroll max
            if (scrollForwardIncrement >= this.scrollMax) {
                return this.scrollMax;
            }
            return scrollForwardIncrement;
        }
        // Scroll backward
        const scrollBackwardIncrement = this.getScrollBackwardStep();
        if (scrollBackwardIncrement <= 0) {
            return 0;
        }
        return scrollBackwardIncrement;
    }
    /**
     * Callback when mouse is still down on the track
     * Incrementally scrolls towards target position until reached
     */
    onTrackOngoingMousedown() {
        const position = this.nextStep();
        return this.scrollTo(position).pipe(
        // If mouse left the track, terminate the stream
        takeUntil(this.pointerOut$), 
        // Keep scrolling until target position is reached
        takeWhile(() => !this.isReached(position)), switchMap(() => this.onTrackOngoingMousedown()));
    }
    /**
     * Returns a flag that determines whether the scroll from the given position is the final step or not
     */
    isReached(position) {
        if (this.scrollDirection === 'forward') {
            return position >= this.scrollMax;
        }
        return position <= 0;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: TrackAdapter, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: TrackAdapter, isStandalone: true, usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: TrackAdapter, decorators: [{
            type: Directive
        }], ctorParameters: () => [] });

class TrackYComponent extends TrackAdapter {
    get contentSize() {
        return this.adapter.contentHeight;
    }
    getThumbStartPosition() {
        return this.control.viewportScrollOffset * this.size / this.contentSize;
    }
    getThumbEndPosition() {
        return (this.control.viewportScrollOffset + this.viewportSize) * this.size / this.contentSize;
    }
    getScrollDirection(position) {
        if (position > this.getThumbEndPosition()) {
            return 'forward';
        }
        else if (position < this.getThumbStartPosition()) {
            return 'backward';
        }
        return null;
    }
    scrollTo(top) {
        return from(this.adapter.scrollTo({ top, duration: this.adapter.trackScrollDuration() }));
    }
    getScrollForwardStep() {
        return this.control.viewportScrollOffset + this.viewportSize;
    }
    getScrollBackwardStep() {
        return this.control.viewportScrollOffset - this.viewportSize;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: TrackYComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.1.2", type: TrackYComponent, isStandalone: true, selector: "scrollbar-track-y", providers: [{ provide: TrackAdapter, useExisting: TrackYComponent }], usesInheritance: true, ngImport: i0, template: '<ng-content/>', isInline: true, styles: [":host{position:relative;width:100%;height:100%;cursor:default;z-index:1;order:2}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: TrackYComponent, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbar-track-y', template: '<ng-content/>', changeDetection: ChangeDetectionStrategy.OnPush, providers: [{ provide: TrackAdapter, useExisting: TrackYComponent }], styles: [":host{position:relative;width:100%;height:100%;cursor:default;z-index:1;order:2}\n"] }]
        }] });
class TrackXComponent extends TrackAdapter {
    get contentSize() {
        return this.adapter.contentWidth;
    }
    getThumbStartPosition;
    getThumbEndPosition;
    getScrollDirection;
    constructor() {
        effect(() => {
            if (this.adapter.dir.valueSignal() === 'rtl') {
                this.getThumbStartPosition = () => {
                    const offset = this.contentSize - this.viewportSize - this.control.viewportScrollOffset;
                    return offset * this.size / this.contentSize;
                };
                this.getThumbEndPosition = () => {
                    const offset = this.contentSize - this.control.viewportScrollOffset;
                    return offset * this.size / this.contentSize;
                };
                this.getScrollDirection = (position) => {
                    if (position < this.getThumbStartPosition()) {
                        return 'forward';
                    }
                    else if (position > this.getThumbEndPosition()) {
                        return 'backward';
                    }
                    return null;
                };
            }
            else {
                this.getThumbStartPosition = () => {
                    return this.control.viewportScrollOffset * this.size / this.contentSize;
                };
                this.getThumbEndPosition = () => {
                    return (this.control.viewportScrollOffset + this.viewportSize) * this.size / this.contentSize;
                };
                this.getScrollDirection = (position) => {
                    if (position > this.getThumbEndPosition()) {
                        return 'forward';
                    }
                    else if (position < this.getThumbStartPosition()) {
                        return 'backward';
                    }
                    return null;
                };
            }
        });
        super();
    }
    scrollTo(start) {
        return from(this.adapter.scrollTo({ start, duration: this.adapter.trackScrollDuration() }));
    }
    getScrollForwardStep() {
        return this.control.viewportScrollOffset + this.viewportSize;
    }
    getScrollBackwardStep() {
        return this.control.viewportScrollOffset - this.viewportSize;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: TrackXComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.1.2", type: TrackXComponent, isStandalone: true, selector: "scrollbar-track-x", providers: [{ provide: TrackAdapter, useExisting: TrackXComponent }], usesInheritance: true, ngImport: i0, template: '<ng-content/>', isInline: true, styles: [":host{position:relative;width:100%;height:100%;cursor:default;z-index:1;order:2}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: TrackXComponent, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbar-track-x', template: '<ng-content/>', changeDetection: ChangeDetectionStrategy.OnPush, providers: [{ provide: TrackAdapter, useExisting: TrackXComponent }], styles: [":host{position:relative;width:100%;height:100%;cursor:default;z-index:1;order:2}\n"] }]
        }], ctorParameters: () => [] });

class ThumbAdapter extends PointerEventsAdapter {
    manager = inject(ScrollbarManager);
    track = inject(TrackAdapter);
    // The animation reference used for enabling the polyfill in Firefox.
    _animation;
    // Returns thumb size
    get size() {
        return this.nativeElement.getBoundingClientRect()[this.control.rectSizeProperty];
    }
    // The maximum space available for scrolling.
    get trackMax() {
        return this.track.size - this.size;
    }
    /**
     * Stream that emits the 'scrollTo' position when a scrollbar thumb element is dragged
     * This function is called by thumb drag event using viewport or scrollbar pointer events
     */
    get pointerEvents() {
        return fromEvent(this.nativeElement, 'pointerdown').pipe(stopPropagation(), preventSelection(this.document), switchMap((e) => {
            let startTrackMax;
            let startScrollMax;
            const dragStart = of(e).pipe(tap(() => {
                // Capture scrollMax and trackMax once
                startTrackMax = this.trackMax;
                startScrollMax = this.control.viewportScrollMax;
                this.setDragging(this.control.axis);
            }));
            const dragging = fromEvent(this.document, 'pointermove').pipe(stopPropagation());
            const dragEnd = fromEvent(this.document, 'pointerup', { capture: true }).pipe(stopPropagation(), enableSelection(this.document), tap(() => this.setDragging('none')));
            return dragStart.pipe(map((startEvent) => startEvent[this.control.offsetProperty]), switchMap((startOffset) => dragging.pipe(map((moveEvent) => moveEvent[this.control.clientProperty]), 
            // Calculate how far the pointer is from the top/left of the scrollbar (minus the dragOffset).
            map((moveClient) => moveClient - this.track.offset), map((trackRelativeOffset) => startScrollMax * (trackRelativeOffset - startOffset) / startTrackMax), tap((scrollPosition) => this.control.instantScrollTo(scrollPosition, startScrollMax)), takeUntil(dragEnd))));
        }));
    }
    constructor() {
        afterRenderEffect({
            earlyRead: () => {
                const script = this.manager.scrollTimelinePolyfill();
                untracked(() => {
                    if (script && !this._animation) {
                        this._animation = startPolyfill(script, this.nativeElement, this.adapter.viewportElement, this.control.axis);
                    }
                });
            }
        });
        super();
    }
    setDragging(value) {
        this.zone.run(() => this.adapter.dragging.set(value));
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ThumbAdapter, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: ThumbAdapter, isStandalone: true, usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ThumbAdapter, decorators: [{
            type: Directive
        }], ctorParameters: () => [] });
function startPolyfill(ScrollTimeline, element, source, axis) {
    return element.animate({
        translate: [
            'var(--_scrollbar-thumb-transform-from)',
            'var(--_scrollbar-thumb-transform-to)'
        ]
    }, {
        fill: 'both',
        easing: 'linear',
        timeline: new ScrollTimeline({ source, axis })
    });
}

class ThumbYComponent extends ThumbAdapter {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ThumbYComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.1.2", type: ThumbYComponent, isStandalone: true, selector: "scrollbar-thumb-y", providers: [{ provide: ThumbAdapter, useExisting: ThumbYComponent }], usesInheritance: true, ngImport: i0, template: '', isInline: true, styles: [":host{--_thumb-size-percent: var(--_fixed-thumb-size-percent, max(var(--_thumb-size-percent-raw), var(--_min-thumb-size-percent)));--_effective-thumb-ratio: calc(var(--_thumb-size-percent) / 100);--_scrollbar-thumb-transform-to-value: calc((100 / var(--_effective-thumb-ratio) - 100) * 1%);box-sizing:border-box;position:absolute;background:var(--_scrollbar-thumb-color);transition:var(--scrollbar-thumb-transition, none);border-radius:var(--scrollbar-thumb-shape, inherit);will-change:transform;animation-name:scrollbarThumbAnimation;animation-duration:1ms;animation-timing-function:linear;-webkit-user-select:none;user-select:none}:host:hover{background:var(--_scrollbar-thumb-hover-color)}:host:active{background:var(--_scrollbar-thumb-active-color)}@keyframes scrollbarThumbAnimation{0%{translate:var(--_scrollbar-thumb-transform-from)}to{translate:var(--_scrollbar-thumb-transform-to)}}\n", ":host{animation-timeline:var(--_animation-timeline-y);display:var(--_vertical-thumb-display);height:calc(var(--_thumb-size-percent) * 1%);width:100%;--_thumb-size-percent-raw: calc(var(--viewport-height) / var(--content-height) * 100);--_min-thumb-size-percent: calc(var(--_scrollbar-thumb-min-size) / var(--viewport-height) * 100);--_fixed-thumb-size-percent: calc(var(--scrollbar-thumb-size) / var(--viewport-height) * 100);--_scrollbar-thumb-transform-from: 0 0;--_scrollbar-thumb-transform-to: 0 var(--_scrollbar-thumb-transform-to-value)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ThumbYComponent, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbar-thumb-y', template: '', changeDetection: ChangeDetectionStrategy.OnPush, providers: [{ provide: ThumbAdapter, useExisting: ThumbYComponent }], styles: [":host{--_thumb-size-percent: var(--_fixed-thumb-size-percent, max(var(--_thumb-size-percent-raw), var(--_min-thumb-size-percent)));--_effective-thumb-ratio: calc(var(--_thumb-size-percent) / 100);--_scrollbar-thumb-transform-to-value: calc((100 / var(--_effective-thumb-ratio) - 100) * 1%);box-sizing:border-box;position:absolute;background:var(--_scrollbar-thumb-color);transition:var(--scrollbar-thumb-transition, none);border-radius:var(--scrollbar-thumb-shape, inherit);will-change:transform;animation-name:scrollbarThumbAnimation;animation-duration:1ms;animation-timing-function:linear;-webkit-user-select:none;user-select:none}:host:hover{background:var(--_scrollbar-thumb-hover-color)}:host:active{background:var(--_scrollbar-thumb-active-color)}@keyframes scrollbarThumbAnimation{0%{translate:var(--_scrollbar-thumb-transform-from)}to{translate:var(--_scrollbar-thumb-transform-to)}}\n", ":host{animation-timeline:var(--_animation-timeline-y);display:var(--_vertical-thumb-display);height:calc(var(--_thumb-size-percent) * 1%);width:100%;--_thumb-size-percent-raw: calc(var(--viewport-height) / var(--content-height) * 100);--_min-thumb-size-percent: calc(var(--_scrollbar-thumb-min-size) / var(--viewport-height) * 100);--_fixed-thumb-size-percent: calc(var(--scrollbar-thumb-size) / var(--viewport-height) * 100);--_scrollbar-thumb-transform-from: 0 0;--_scrollbar-thumb-transform-to: 0 var(--_scrollbar-thumb-transform-to-value)}\n"] }]
        }] });
class ThumbXComponent extends ThumbAdapter {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ThumbXComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.1.2", type: ThumbXComponent, isStandalone: true, selector: "scrollbar-thumb-x", providers: [{ provide: ThumbAdapter, useExisting: ThumbXComponent }], usesInheritance: true, ngImport: i0, template: '', isInline: true, styles: [":host{--_thumb-size-percent: var(--_fixed-thumb-size-percent, max(var(--_thumb-size-percent-raw), var(--_min-thumb-size-percent)));--_effective-thumb-ratio: calc(var(--_thumb-size-percent) / 100);--_scrollbar-thumb-transform-to-value: calc((100 / var(--_effective-thumb-ratio) - 100) * 1%);box-sizing:border-box;position:absolute;background:var(--_scrollbar-thumb-color);transition:var(--scrollbar-thumb-transition, none);border-radius:var(--scrollbar-thumb-shape, inherit);will-change:transform;animation-name:scrollbarThumbAnimation;animation-duration:1ms;animation-timing-function:linear;-webkit-user-select:none;user-select:none}:host:hover{background:var(--_scrollbar-thumb-hover-color)}:host:active{background:var(--_scrollbar-thumb-active-color)}@keyframes scrollbarThumbAnimation{0%{translate:var(--_scrollbar-thumb-transform-from)}to{translate:var(--_scrollbar-thumb-transform-to)}}\n", ":host{animation-timeline:var(--_animation-timeline-x);display:var(--_horizontal-thumb-display);height:100%;width:calc(var(--_thumb-size-percent) * 1%);--_thumb-size-percent-raw: calc(var(--viewport-width) / var(--content-width) * 100);--_min-thumb-size-percent: calc(var(--_scrollbar-thumb-min-size) / var(--viewport-width) * 100);--_fixed-thumb-size-percent: calc(var(--scrollbar-thumb-size) / var(--viewport-width) * 100);--_scrollbar-thumb-transform-from: 0;--_scrollbar-thumb-transform-to: var(--_scrollbar-thumb-transform-to-value)}:host[dir=rtl]{animation-name:scrollbarThumbRTLAnimation;will-change:right;--_scrollbar-thumb-transform-to: calc(var(--_scrollbar-thumb-transform-to-value) * -1)}@keyframes scrollbarThumbRTLAnimation{0%{right:var(--_scrollbar-thumb-transform-from)}to{right:calc((var(--_effective-thumb-ratio) - 100 + var(--_effective-thumb-ratio) * 100) * -1%)}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ThumbXComponent, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbar-thumb-x', template: '', changeDetection: ChangeDetectionStrategy.OnPush, providers: [{ provide: ThumbAdapter, useExisting: ThumbXComponent }], styles: [":host{--_thumb-size-percent: var(--_fixed-thumb-size-percent, max(var(--_thumb-size-percent-raw), var(--_min-thumb-size-percent)));--_effective-thumb-ratio: calc(var(--_thumb-size-percent) / 100);--_scrollbar-thumb-transform-to-value: calc((100 / var(--_effective-thumb-ratio) - 100) * 1%);box-sizing:border-box;position:absolute;background:var(--_scrollbar-thumb-color);transition:var(--scrollbar-thumb-transition, none);border-radius:var(--scrollbar-thumb-shape, inherit);will-change:transform;animation-name:scrollbarThumbAnimation;animation-duration:1ms;animation-timing-function:linear;-webkit-user-select:none;user-select:none}:host:hover{background:var(--_scrollbar-thumb-hover-color)}:host:active{background:var(--_scrollbar-thumb-active-color)}@keyframes scrollbarThumbAnimation{0%{translate:var(--_scrollbar-thumb-transform-from)}to{translate:var(--_scrollbar-thumb-transform-to)}}\n", ":host{animation-timeline:var(--_animation-timeline-x);display:var(--_horizontal-thumb-display);height:100%;width:calc(var(--_thumb-size-percent) * 1%);--_thumb-size-percent-raw: calc(var(--viewport-width) / var(--content-width) * 100);--_min-thumb-size-percent: calc(var(--_scrollbar-thumb-min-size) / var(--viewport-width) * 100);--_fixed-thumb-size-percent: calc(var(--scrollbar-thumb-size) / var(--viewport-width) * 100);--_scrollbar-thumb-transform-from: 0;--_scrollbar-thumb-transform-to: var(--_scrollbar-thumb-transform-to-value)}:host[dir=rtl]{animation-name:scrollbarThumbRTLAnimation;will-change:right;--_scrollbar-thumb-transform-to: calc(var(--_scrollbar-thumb-transform-to-value) * -1)}@keyframes scrollbarThumbRTLAnimation{0%{right:var(--_scrollbar-thumb-transform-from)}to{right:calc((var(--_effective-thumb-ratio) - 100 + var(--_effective-thumb-ratio) * 100) * -1%)}}\n"] }]
        }] });

// canScroll function can work for y-axis and x-axis for both LTR and RTL directions
const canScrollFunc = {
    forward: (scrollOffset, scrollMax) => scrollOffset < scrollMax,
    backward: (scrollOffset) => scrollOffset > 0
};
const scrollStepFunc = {
    forward: (scrollBy, offset) => offset + scrollBy,
    backward: (scrollBy, offset) => offset - scrollBy
};
const horizontalScrollStepFunc = {
    rtl: {
        forward: (scrollBy, offset, scrollMax) => scrollMax - offset - scrollBy,
        backward: (scrollBy, offset, scrollMax) => scrollMax - offset + scrollBy
    },
    ltr: scrollStepFunc
};
class ScrollbarButton extends PointerEventsAdapter {
    scrollbarButton = input.required(...(ngDevMode ? [{ debugName: "scrollbarButton" }] : []));
    scrollDirection = input.required(...(ngDevMode ? [{ debugName: "scrollDirection" }] : []));
    afterFirstClickDelay = 120;
    firstClickDuration = 100;
    scrollBy = 50;
    onGoingScrollBy = 12;
    nextStep;
    canScroll;
    get pointerEvents() {
        const pointerDown$ = fromEvent(this.nativeElement, 'pointerdown').pipe(stopPropagation(), preventSelection(this.document));
        const pointerUp$ = fromEvent(this.document, 'pointerup', { passive: true }).pipe(enableSelection(this.document));
        const pointerLeave$ = fromEvent(this.nativeElement, 'pointerleave', { passive: true });
        // Combine pointerup and pointerleave events into one stream
        const pointerUpOrLeave$ = merge(pointerUp$, pointerLeave$);
        return pointerDown$.pipe(switchMap(() => this.firstScrollStep().pipe(delay(this.afterFirstClickDelay), switchMap(() => this.onOngoingPointerdown()), takeUntil(pointerUpOrLeave$))));
    }
    constructor() {
        effect(() => {
            const scrollDirection = this.scrollDirection();
            const dir = this.adapter.dir.valueSignal();
            untracked(() => {
                // Get the canScroll function according to the scroll direction (forward/backward)
                this.canScroll = canScrollFunc[scrollDirection];
                if (this.control.axis === 'x') {
                    // Get the nextStep function according to the scroll direction (forward/backward) and layout direction (LTR/RTL)
                    this.nextStep = horizontalScrollStepFunc[dir][scrollDirection];
                }
                else {
                    // Get the nextStep function according to the scroll direction (forward/backward)
                    this.nextStep = scrollStepFunc[scrollDirection];
                }
            });
        });
        super();
    }
    firstScrollStep() {
        const value = this.nextStep(this.scrollBy, this.control.viewportScrollOffset, this.control.viewportScrollMax);
        return this.control.scrollTo(value, this.firstClickDuration);
    }
    onGoingScrollStep() {
        const scrollMax = this.control.viewportScrollMax;
        const value = this.nextStep(this.onGoingScrollBy, this.control.viewportScrollOffset, scrollMax);
        this.control.instantScrollTo(value, scrollMax);
    }
    onOngoingPointerdown() {
        return interval(0, animationFrameScheduler).pipe(takeWhile(() => this.canScroll(this.control.viewportScrollOffset, this.control.viewportScrollMax)), tap(() => this.onGoingScrollStep()));
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarButton, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.1.2", type: ScrollbarButton, isStandalone: true, selector: "button[scrollbarButton]", inputs: { scrollbarButton: { classPropertyName: "scrollbarButton", publicName: "scrollbarButton", isSignal: true, isRequired: true, transformFunction: null }, scrollDirection: { classPropertyName: "scrollDirection", publicName: "scrollDirection", isSignal: true, isRequired: true, transformFunction: null } }, usesInheritance: true, ngImport: i0, template: "<div class=\"ng-scrollbar-button-icon\">\r\n  <svg viewBox=\"0 0 512 512\"\r\n       xmlns=\"http://www.w3.org/2000/svg\">\r\n    <path\r\n      d=\"M413.1,327.3l-1.8-2.1l-136-156.5c-4.6-5.3-11.5-8.6-19.2-8.6c-7.7,0-14.6,3.4-19.2,8.6L101,324.9l-2.3,2.6  C97,330,96,333,96,336.2c0,8.7,7.4,15.8,16.6,15.8v0h286.8v0c9.2,0,16.6-7.1,16.6-15.8C416,332.9,414.9,329.8,413.1,327.3z\"/>\r\n  </svg>\r\n</div>\r\n", styles: [":host{position:relative;border:none;margin:0;padding:0;border-radius:var(--scrollbar-button-shape);appearance:none;background:var(--_scrollbar-button-color);flex:0 0 var(--scrollbar-button-size, var(--_scrollbar-track-thickness-px));block-size:100%}:host svg{width:100%;height:100%;fill:var(--_scrollbar-button-fill)}:host:hover{background:var(--_scrollbar-button-hover-color)}:host:hover svg{fill:var(--_scrollbar-button-hover-fill)}:host:active{background:var(--_scrollbar-button-active-color)}:host:active svg{fill:var(--_scrollbar-button-active-fill)}:host:disabled{background:var(--_scrollbar-button-inactive-color)}:host:disabled svg{fill:var(--_scrollbar-button-inactive-fill)}:host[scrollbarButton=top]{--_button-rotate: var(--_button-icon-rotate-top)}:host[scrollbarButton=bottom]{--_button-rotate: var(--_button-icon-rotate-bottom)}:host[scrollbarButton=start]{--_button-rotate: var(--_button-icon-rotate-start)}:host[scrollbarButton=end]{--_button-rotate: var(--_button-icon-rotate-end)}:host:is([scrollbarButton=top],[scrollbarButton=start]){order:1}:host:is([scrollbarButton=bottom],[scrollbarButton=end]){order:3}.ng-scrollbar-button-icon{rotate:var(--_button-rotate);display:flex;place-content:center;place-items:center;width:100%;height:100%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarButton, decorators: [{
            type: Component,
            args: [{ selector: 'button[scrollbarButton]', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"ng-scrollbar-button-icon\">\r\n  <svg viewBox=\"0 0 512 512\"\r\n       xmlns=\"http://www.w3.org/2000/svg\">\r\n    <path\r\n      d=\"M413.1,327.3l-1.8-2.1l-136-156.5c-4.6-5.3-11.5-8.6-19.2-8.6c-7.7,0-14.6,3.4-19.2,8.6L101,324.9l-2.3,2.6  C97,330,96,333,96,336.2c0,8.7,7.4,15.8,16.6,15.8v0h286.8v0c9.2,0,16.6-7.1,16.6-15.8C416,332.9,414.9,329.8,413.1,327.3z\"/>\r\n  </svg>\r\n</div>\r\n", styles: [":host{position:relative;border:none;margin:0;padding:0;border-radius:var(--scrollbar-button-shape);appearance:none;background:var(--_scrollbar-button-color);flex:0 0 var(--scrollbar-button-size, var(--_scrollbar-track-thickness-px));block-size:100%}:host svg{width:100%;height:100%;fill:var(--_scrollbar-button-fill)}:host:hover{background:var(--_scrollbar-button-hover-color)}:host:hover svg{fill:var(--_scrollbar-button-hover-fill)}:host:active{background:var(--_scrollbar-button-active-color)}:host:active svg{fill:var(--_scrollbar-button-active-fill)}:host:disabled{background:var(--_scrollbar-button-inactive-color)}:host:disabled svg{fill:var(--_scrollbar-button-inactive-fill)}:host[scrollbarButton=top]{--_button-rotate: var(--_button-icon-rotate-top)}:host[scrollbarButton=bottom]{--_button-rotate: var(--_button-icon-rotate-bottom)}:host[scrollbarButton=start]{--_button-rotate: var(--_button-icon-rotate-start)}:host[scrollbarButton=end]{--_button-rotate: var(--_button-icon-rotate-end)}:host:is([scrollbarButton=top],[scrollbarButton=start]){order:1}:host:is([scrollbarButton=bottom],[scrollbarButton=end]){order:3}.ng-scrollbar-button-icon{rotate:var(--_button-rotate);display:flex;place-content:center;place-items:center;width:100%;height:100%}\n"] }]
        }], ctorParameters: () => [], propDecorators: { scrollbarButton: [{ type: i0.Input, args: [{ isSignal: true, alias: "scrollbarButton", required: true }] }], scrollDirection: [{ type: i0.Input, args: [{ isSignal: true, alias: "scrollDirection", required: true }] }] } });

class ScrollbarY extends ScrollbarAdapter {
    rectOffsetProperty = 'top';
    rectSizeProperty = 'height';
    sizeProperty = 'clientHeight';
    clientProperty = 'clientY';
    offsetProperty = 'offsetY';
    axis = 'y';
    get viewportScrollMax() {
        return this.cmp.scrollMaxY;
    }
    get viewportScrollOffset() {
        return this.cmp.scrollTop;
    }
    scrollTo(top, duration) {
        return from(this.cmp.scrollTo({ top, duration }));
    }
    instantScrollTo(value) {
        this.cmp.scrollYTo(value);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarY, deps: null, target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.1.2", type: ScrollbarY, isStandalone: true, selector: "scrollbar-y", providers: [
            { provide: SCROLLBAR_CONTROL, useExisting: ScrollbarY }
        ], usesInheritance: true, ngImport: i0, template: `
    <div class="ng-scrollbar-track-wrapper"
         [class.ng-scrollbar-hover]="!cmp.hoverOffset()">
      <scrollbar-track-y class="ng-scrollbar-track {{ cmp.trackClass() }}">
        <div class="track-inner">
          <scrollbar-thumb-y class="ng-scrollbar-thumb {{ cmp.thumbClass() }}"/>
        </div>
      </scrollbar-track-y>
      @if (cmp.withButtons()) {
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="top"
                scrollDirection="backward"
                [disabled]="!cmp.isVerticallyScrollable()"></button>
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="bottom"
                scrollDirection="forward"
                [disabled]="!cmp.isVerticallyScrollable()"></button>
      }
    </div>
  `, isInline: true, styles: [":host{display:block;height:var(--_scrollbar-wrapper-height);width:var(--_scrollbar-wrapper-width);position:sticky;z-index:100;opacity:var(--_scrollbar-hover-opacity);transition:var(--_scrollbar-opacity-transition);background:var(--scrollbar-container-color, transparent);border-radius:var(--scrollbar-container-shape, 0px)}.ng-scrollbar-track-wrapper{touch-action:none;-webkit-user-select:none;user-select:none;gap:var(--scrollbar-button-gap, 0px);inset-block-start:var(--_scrollbar-track-inset-block-start);inset-block-end:var(--_scrollbar-track-inset-block-end);inset-inline-start:var(--_scrollbar-track-inset-inline-start);inset-inline-end:var(--_scrollbar-track-inset-inline-end);transition:var(--scrollbar-track-wrapper-transition, width 60ms linear, height 60ms linear);position:absolute;overflow:hidden;display:flex;place-items:center}.track-inner{position:absolute;border-radius:var(--scrollbar-track-shape, 0);background:var(--scrollbar-track-color, rgba(128, 128, 128, .5))}\n", ":host{--_scrollbar-wrapper-height: calc(var(--viewport-height) * 1px);--_scrollbar-wrapper-width: var(--_scrollbar-total-thickness-px);--_scrollbar-track-inset-block-start: var(--_vertical-inset-block-start);--_scrollbar-track-inset-block-end: var(--_vertical-inset-block-end);--_scrollbar-track-inset-inline-end: var(--_vertical-inset-inline-end);--_scrollbar-track-inset-inline-start: var(--_vertical-inset-inline-start);--_button-icon-rotate-top: 0deg;--_button-icon-rotate-bottom: 180deg;grid-area:scrollbar-y;pointer-events:var(--_scrollbar-y-pointer-events);inset-block-start:0;inset-block-end:0;inset-inline-start:var(--_scrollbar-wrapper-y-start);inset-inline-end:var(--_scrollbar-wrapper-y-end)}:host.ng-scrollbar-hover:hover,.ng-scrollbar-hover:hover{--_track-y-thickness-px: var(--_scrollbar-hover-thickness-px)}:host.ng-scrollbar-hover:active,.ng-scrollbar-hover:active{--_track-y-thickness-px: var(--_scrollbar-active-thickness-px)}.ng-scrollbar-track-wrapper{width:var(--_track-y-thickness-px);flex-direction:column}.track-inner{inset-inline-start:var(--scrollbar-track-offset, 0px);inset-inline-end:var(--scrollbar-track-offset, 0px);inset-block-start:0;inset-block-end:0}\n"], dependencies: [{ kind: "component", type: TrackYComponent, selector: "scrollbar-track-y" }, { kind: "component", type: ThumbYComponent, selector: "scrollbar-thumb-y" }, { kind: "component", type: ScrollbarButton, selector: "button[scrollbarButton]", inputs: ["scrollbarButton", "scrollDirection"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarY, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbar-y', template: `
    <div class="ng-scrollbar-track-wrapper"
         [class.ng-scrollbar-hover]="!cmp.hoverOffset()">
      <scrollbar-track-y class="ng-scrollbar-track {{ cmp.trackClass() }}">
        <div class="track-inner">
          <scrollbar-thumb-y class="ng-scrollbar-thumb {{ cmp.thumbClass() }}"/>
        </div>
      </scrollbar-track-y>
      @if (cmp.withButtons()) {
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="top"
                scrollDirection="backward"
                [disabled]="!cmp.isVerticallyScrollable()"></button>
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="bottom"
                scrollDirection="forward"
                [disabled]="!cmp.isVerticallyScrollable()"></button>
      }
    </div>
  `, imports: [TrackYComponent, ThumbYComponent, ScrollbarButton], providers: [
                        { provide: SCROLLBAR_CONTROL, useExisting: ScrollbarY }
                    ], changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{display:block;height:var(--_scrollbar-wrapper-height);width:var(--_scrollbar-wrapper-width);position:sticky;z-index:100;opacity:var(--_scrollbar-hover-opacity);transition:var(--_scrollbar-opacity-transition);background:var(--scrollbar-container-color, transparent);border-radius:var(--scrollbar-container-shape, 0px)}.ng-scrollbar-track-wrapper{touch-action:none;-webkit-user-select:none;user-select:none;gap:var(--scrollbar-button-gap, 0px);inset-block-start:var(--_scrollbar-track-inset-block-start);inset-block-end:var(--_scrollbar-track-inset-block-end);inset-inline-start:var(--_scrollbar-track-inset-inline-start);inset-inline-end:var(--_scrollbar-track-inset-inline-end);transition:var(--scrollbar-track-wrapper-transition, width 60ms linear, height 60ms linear);position:absolute;overflow:hidden;display:flex;place-items:center}.track-inner{position:absolute;border-radius:var(--scrollbar-track-shape, 0);background:var(--scrollbar-track-color, rgba(128, 128, 128, .5))}\n", ":host{--_scrollbar-wrapper-height: calc(var(--viewport-height) * 1px);--_scrollbar-wrapper-width: var(--_scrollbar-total-thickness-px);--_scrollbar-track-inset-block-start: var(--_vertical-inset-block-start);--_scrollbar-track-inset-block-end: var(--_vertical-inset-block-end);--_scrollbar-track-inset-inline-end: var(--_vertical-inset-inline-end);--_scrollbar-track-inset-inline-start: var(--_vertical-inset-inline-start);--_button-icon-rotate-top: 0deg;--_button-icon-rotate-bottom: 180deg;grid-area:scrollbar-y;pointer-events:var(--_scrollbar-y-pointer-events);inset-block-start:0;inset-block-end:0;inset-inline-start:var(--_scrollbar-wrapper-y-start);inset-inline-end:var(--_scrollbar-wrapper-y-end)}:host.ng-scrollbar-hover:hover,.ng-scrollbar-hover:hover{--_track-y-thickness-px: var(--_scrollbar-hover-thickness-px)}:host.ng-scrollbar-hover:active,.ng-scrollbar-hover:active{--_track-y-thickness-px: var(--_scrollbar-active-thickness-px)}.ng-scrollbar-track-wrapper{width:var(--_track-y-thickness-px);flex-direction:column}.track-inner{inset-inline-start:var(--scrollbar-track-offset, 0px);inset-inline-end:var(--scrollbar-track-offset, 0px);inset-block-start:0;inset-block-end:0}\n"] }]
        }] });
class ScrollbarX extends ScrollbarAdapter {
    manager = inject(ScrollbarManager);
    rectOffsetProperty = 'left';
    rectSizeProperty = 'width';
    sizeProperty = 'clientWidth';
    clientProperty = 'clientX';
    offsetProperty = 'offsetX';
    axis = 'x';
    // Handle dragging position (Support LTR and RTL directions for the horizontal scrollbar)
    handlePosition;
    get viewportScrollMax() {
        return this.cmp.scrollMaxX;
    }
    get viewportScrollOffset() {
        // Keep scrollLeft value positive for horizontal scrollbar
        return Math.abs(this.cmp.scrollLeft);
    }
    constructor() {
        effect(() => {
            if (this.cmp.dir.valueSignal() === 'rtl') {
                this.handlePosition = (position, scrollMax) => -(scrollMax - position);
            }
            else {
                this.handlePosition = (position) => position;
            }
        });
        super();
    }
    scrollTo(left, duration) {
        return from(this.cmp.scrollTo({ left, duration }));
    }
    instantScrollTo(value, scrollMax) {
        this.cmp.scrollXTo(this.handlePosition(value, scrollMax));
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarX, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.1.2", type: ScrollbarX, isStandalone: true, selector: "scrollbar-x", host: { properties: { "attr.dir": "cmp.dir.valueSignal()" } }, providers: [
            { provide: SCROLLBAR_CONTROL, useExisting: ScrollbarX }
        ], usesInheritance: true, ngImport: i0, template: `
    <div class="ng-scrollbar-track-wrapper"
         [class.ng-scrollbar-hover]="!cmp.hoverOffset()">
      <scrollbar-track-x class="ng-scrollbar-track {{ cmp.trackClass() }}">
        <div class="track-inner">
          <scrollbar-thumb-x class="ng-scrollbar-thumb {{ cmp.thumbClass() }}" [attr.dir]="cmp.dir.valueSignal()"/>
        </div>
      </scrollbar-track-x>
      @if (cmp.withButtons()) {
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="start"
                scrollDirection="backward"
                [disabled]="!cmp.isHorizontallyScrollable()"></button>
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="end"
                scrollDirection="forward"
                [disabled]="!cmp.isHorizontallyScrollable()"></button>
      }
    </div>
  `, isInline: true, styles: [":host{display:block;height:var(--_scrollbar-wrapper-height);width:var(--_scrollbar-wrapper-width);position:sticky;z-index:100;opacity:var(--_scrollbar-hover-opacity);transition:var(--_scrollbar-opacity-transition);background:var(--scrollbar-container-color, transparent);border-radius:var(--scrollbar-container-shape, 0px)}.ng-scrollbar-track-wrapper{touch-action:none;-webkit-user-select:none;user-select:none;gap:var(--scrollbar-button-gap, 0px);inset-block-start:var(--_scrollbar-track-inset-block-start);inset-block-end:var(--_scrollbar-track-inset-block-end);inset-inline-start:var(--_scrollbar-track-inset-inline-start);inset-inline-end:var(--_scrollbar-track-inset-inline-end);transition:var(--scrollbar-track-wrapper-transition, width 60ms linear, height 60ms linear);position:absolute;overflow:hidden;display:flex;place-items:center}.track-inner{position:absolute;border-radius:var(--scrollbar-track-shape, 0);background:var(--scrollbar-track-color, rgba(128, 128, 128, .5))}\n", ":host{--_scrollbar-wrapper-height: var(--_scrollbar-total-thickness-px);--_scrollbar-wrapper-width: calc(var(--viewport-width) * 1px);--_scrollbar-track-inset-block-start: var(--_horizontal-block-start);--_scrollbar-track-inset-block-end: var(--_horizontal-block-end);--_scrollbar-track-inset-inline-end: var(--_horizontal-inline-end);--_scrollbar-track-inset-inline-start: var(--_horizontal-inline-start);--_button-icon-rotate-start: -90deg;--_button-icon-rotate-end: 90deg}:host[dir=rtl]{--_button-icon-rotate-start: 90deg;--_button-icon-rotate-end: -90deg}:host{grid-area:scrollbar-x;pointer-events:var(--_scrollbar-x-pointer-events);inset-block-start:var(--_scrollbar-wrapper-x-top);inset-block-end:var(--_scrollbar-wrapper-x-bottom);inset-inline-start:0;inset-inline-end:0}:host.ng-scrollbar-hover:hover,.ng-scrollbar-hover:hover{--_track-x-thickness-px: var(--_scrollbar-hover-thickness-px)}:host.ng-scrollbar-hover:active,.ng-scrollbar-hover:active{--_track-x-thickness-px: var(--_scrollbar-active-thickness-px)}.ng-scrollbar-track-wrapper{height:var(--_track-x-thickness-px);flex-direction:row}.track-inner{inset-inline-start:0;inset-inline-end:0;inset-block-start:var(--scrollbar-track-offset, 0px);inset-block-end:var(--scrollbar-track-offset, 0px)}\n"], dependencies: [{ kind: "component", type: TrackXComponent, selector: "scrollbar-track-x" }, { kind: "component", type: ThumbXComponent, selector: "scrollbar-thumb-x" }, { kind: "component", type: ScrollbarButton, selector: "button[scrollbarButton]", inputs: ["scrollbarButton", "scrollDirection"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollbarX, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbar-x', template: `
    <div class="ng-scrollbar-track-wrapper"
         [class.ng-scrollbar-hover]="!cmp.hoverOffset()">
      <scrollbar-track-x class="ng-scrollbar-track {{ cmp.trackClass() }}">
        <div class="track-inner">
          <scrollbar-thumb-x class="ng-scrollbar-thumb {{ cmp.thumbClass() }}" [attr.dir]="cmp.dir.valueSignal()"/>
        </div>
      </scrollbar-track-x>
      @if (cmp.withButtons()) {
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="start"
                scrollDirection="backward"
                [disabled]="!cmp.isHorizontallyScrollable()"></button>
        <!-- eslint-disable-next-line @angular-eslint/template/elements-content -->
        <button class="ng-scrollbar-button {{ cmp.buttonClass() }}"
                scrollbarButton="end"
                scrollDirection="forward"
                [disabled]="!cmp.isHorizontallyScrollable()"></button>
      }
    </div>
  `, imports: [TrackXComponent, ThumbXComponent, ScrollbarButton], providers: [
                        { provide: SCROLLBAR_CONTROL, useExisting: ScrollbarX }
                    ], host: {
                        '[attr.dir]': 'cmp.dir.valueSignal()'
                    }, changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{display:block;height:var(--_scrollbar-wrapper-height);width:var(--_scrollbar-wrapper-width);position:sticky;z-index:100;opacity:var(--_scrollbar-hover-opacity);transition:var(--_scrollbar-opacity-transition);background:var(--scrollbar-container-color, transparent);border-radius:var(--scrollbar-container-shape, 0px)}.ng-scrollbar-track-wrapper{touch-action:none;-webkit-user-select:none;user-select:none;gap:var(--scrollbar-button-gap, 0px);inset-block-start:var(--_scrollbar-track-inset-block-start);inset-block-end:var(--_scrollbar-track-inset-block-end);inset-inline-start:var(--_scrollbar-track-inset-inline-start);inset-inline-end:var(--_scrollbar-track-inset-inline-end);transition:var(--scrollbar-track-wrapper-transition, width 60ms linear, height 60ms linear);position:absolute;overflow:hidden;display:flex;place-items:center}.track-inner{position:absolute;border-radius:var(--scrollbar-track-shape, 0);background:var(--scrollbar-track-color, rgba(128, 128, 128, .5))}\n", ":host{--_scrollbar-wrapper-height: var(--_scrollbar-total-thickness-px);--_scrollbar-wrapper-width: calc(var(--viewport-width) * 1px);--_scrollbar-track-inset-block-start: var(--_horizontal-block-start);--_scrollbar-track-inset-block-end: var(--_horizontal-block-end);--_scrollbar-track-inset-inline-end: var(--_horizontal-inline-end);--_scrollbar-track-inset-inline-start: var(--_horizontal-inline-start);--_button-icon-rotate-start: -90deg;--_button-icon-rotate-end: 90deg}:host[dir=rtl]{--_button-icon-rotate-start: 90deg;--_button-icon-rotate-end: -90deg}:host{grid-area:scrollbar-x;pointer-events:var(--_scrollbar-x-pointer-events);inset-block-start:var(--_scrollbar-wrapper-x-top);inset-block-end:var(--_scrollbar-wrapper-x-bottom);inset-inline-start:0;inset-inline-end:0}:host.ng-scrollbar-hover:hover,.ng-scrollbar-hover:hover{--_track-x-thickness-px: var(--_scrollbar-hover-thickness-px)}:host.ng-scrollbar-hover:active,.ng-scrollbar-hover:active{--_track-x-thickness-px: var(--_scrollbar-active-thickness-px)}.ng-scrollbar-track-wrapper{height:var(--_track-x-thickness-px);flex-direction:row}.track-inner{inset-inline-start:0;inset-inline-end:0;inset-block-start:var(--scrollbar-track-offset, 0px);inset-block-end:var(--scrollbar-track-offset, 0px)}\n"] }]
        }], ctorParameters: () => [] });

class Scrollbars {
    adapter = inject(ViewportAdapter);
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: Scrollbars, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.1.2", type: Scrollbars, isStandalone: true, selector: "scrollbars", ngImport: i0, template: `
    @if (adapter.verticalUsed()) {
      <scrollbar-y/>
    }
    @if (adapter.horizontalUsed()) {
      <scrollbar-x/>
    }
  `, isInline: true, styles: [":host{pointer-events:none;position:absolute;inset:0;display:grid;grid-template:var(--_scrollbar-grid-template)}\n"], dependencies: [{ kind: "component", type: ScrollbarX, selector: "scrollbar-x" }, { kind: "component", type: ScrollbarY, selector: "scrollbar-y" }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: Scrollbars, decorators: [{
            type: Component,
            args: [{ selector: 'scrollbars', changeDetection: ChangeDetectionStrategy.OnPush, imports: [ScrollbarX, ScrollbarY], template: `
    @if (adapter.verticalUsed()) {
      <scrollbar-y/>
    }
    @if (adapter.horizontalUsed()) {
      <scrollbar-x/>
    }
  `, styles: [":host{pointer-events:none;position:absolute;inset:0;display:grid;grid-template:var(--_scrollbar-grid-template)}\n"] }]
        }] });

class NgScrollbarCore {
    zone = inject(NgZone);
    platform = inject(Platform);
    sharedResizeObserver = inject(SharedResizeObserver);
    /** Viewport adapter instance */
    adapter = inject(ViewportAdapter);
    /** Viewport native element */
    nativeElement = inject((ElementRef)).nativeElement;
    /** A flag that indicates if the platform is mobile */
    isMobile = this.platform.IOS || this.platform.ANDROID;
    constructor() {
        let resizeSub$;
        let hasInitialized;
        afterRenderEffect({
            earlyRead: (onCleanup) => {
                const disableSensor = this.adapter.disableSensor();
                const throttleDuration = this.adapter.sensorThrottleTime();
                const viewportInit = this.adapter.initialized();
                untracked(() => {
                    if (viewportInit) {
                        // If resize sensor is disabled, update manually the first time
                        if (disableSensor) {
                            requestAnimationFrame(() => this.update(ScrollbarUpdateReason.AfterInit));
                        }
                        else {
                            // Observe size changes for viewport and content wrapper
                            this.zone.runOutsideAngular(() => {
                                resizeSub$ = getThrottledStream(combineLatest([
                                    this.sharedResizeObserver.observe(this.adapter.viewportElement),
                                    this.sharedResizeObserver.observe(this.adapter.contentWrapperElement)
                                ]), throttleDuration).subscribe(() => {
                                    // After deep investigation, it appears that setting the dimension directly from the element properties
                                    // is much faster than to set them from resize callback values
                                    this.zone.run(() => {
                                        this.updateDimensions();
                                        if (hasInitialized) {
                                            this.adapter.afterUpdate.emit();
                                        }
                                        else {
                                            this.adapter.afterInit.emit();
                                        }
                                        hasInitialized = true;
                                    });
                                });
                            });
                        }
                    }
                    onCleanup(() => resizeSub$?.unsubscribe());
                });
            }
        });
    }
    /**
     * Manual update
     */
    update(reason) {
        this.updateDimensions();
        if (reason === ScrollbarUpdateReason.AfterInit) {
            this.adapter.afterInit.emit();
        }
        else {
            this.adapter.afterUpdate.emit();
        }
    }
    updateDimensions() {
        this.adapter.viewportDimension.set({ width: this.adapter.clientWidth, height: this.adapter.clientHeight });
        this.adapter.contentDimension.set({ width: this.adapter.contentWidth, height: this.adapter.contentHeight });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarCore, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: NgScrollbarCore, isStandalone: true, host: { properties: { "class.ng-scroll-viewport": "true", "attr.mobile": "isMobile", "attr.dir": "adapter.dir.valueSignal()", "attr.dragging": "adapter.dragging()", "attr.position": "adapter.position()", "attr.appearance": "adapter.appearance()", "attr.visibility": "adapter.visibility()", "attr.orientation": "adapter.orientation()", "attr.disableInteraction": "adapter.disableInteraction()", "attr.verticalUsed": "adapter.verticalUsed()", "attr.horizontalUsed": "adapter.horizontalUsed()", "attr.isVerticallyScrollable": "adapter.isVerticallyScrollable()", "attr.isHorizontallyScrollable": "adapter.isHorizontallyScrollable()", "style.--content-height": "adapter.contentDimension().height", "style.--content-width": "adapter.contentDimension().width", "style.--viewport-height": "adapter.viewportDimension().height", "style.--viewport-width": "adapter.viewportDimension().width" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarCore, decorators: [{
            type: Directive,
            args: [{
                    host: {
                        '[class.ng-scroll-viewport]': 'true',
                        '[attr.mobile]': 'isMobile',
                        '[attr.dir]': 'adapter.dir.valueSignal()',
                        '[attr.dragging]': 'adapter.dragging()',
                        '[attr.position]': 'adapter.position()',
                        '[attr.appearance]': 'adapter.appearance()',
                        '[attr.visibility]': 'adapter.visibility()',
                        '[attr.orientation]': 'adapter.orientation()',
                        '[attr.disableInteraction]': 'adapter.disableInteraction()',
                        '[attr.verticalUsed]': 'adapter.verticalUsed()',
                        '[attr.horizontalUsed]': 'adapter.horizontalUsed()',
                        '[attr.isVerticallyScrollable]': 'adapter.isVerticallyScrollable()',
                        '[attr.isHorizontallyScrollable]': 'adapter.isHorizontallyScrollable()',
                        '[style.--content-height]': 'adapter.contentDimension().height',
                        '[style.--content-width]': 'adapter.contentDimension().width',
                        '[style.--viewport-height]': 'adapter.viewportDimension().height',
                        '[style.--viewport-width]': 'adapter.viewportDimension().width'
                    }
                }]
        }], ctorParameters: () => [] });

class ScrollViewport extends NgScrollbarCore {
    renderer = inject(Renderer2);
    appRef = inject(ApplicationRef);
    injector = inject(Injector);
    scrollbarsRef;
    contentWrapperRef;
    /**
     * The element that wraps the content inside the viewport.
     */
    contentWrapperElement;
    /**
     * The spacer element used by virtual scroll component.
     */
    spacerElement;
    constructor() {
        afterNextRender({
            write: () => {
                // Create scroll content
                this.createContentWrapper(this.contentWrapperElement);
                // Attach scrollbars
                this.attachScrollbars(this.spacerElement || this.contentWrapperElement);
                // Initialize viewport
                this.adapter.init(this.nativeElement, this.contentWrapperElement, this.spacerElement);
                // If spaceElement is provided, add the appropriate class
                this.spacerElement?.classList.add(ViewportClasses.Spacer);
            }
        });
        super();
    }
    ngOnDestroy() {
        this.adapter.reset();
        if (this.scrollbarsRef) {
            this.appRef.detachView(this.scrollbarsRef.hostView);
            this.scrollbarsRef.destroy();
            this.scrollbarsRef = null;
        }
        if (this.contentWrapperRef) {
            this.appRef.detachView(this.contentWrapperRef.hostView);
            this.contentWrapperRef.destroy();
            this.contentWrapperRef = null;
        }
    }
    createContentWrapper(hostElement) {
        if (hostElement) {
            // Attach a content wrapper component to a given host element
            this.contentWrapperRef = createComponent(ScrollContent, {
                hostElement,
                elementInjector: this.injector,
                environmentInjector: this.appRef.injector,
                projectableNodes: [Array.from(hostElement.childNodes)]
            });
        }
        else {
            this.contentWrapperRef = createComponent(ScrollContent, {
                elementInjector: this.injector,
                environmentInjector: this.appRef.injector,
                projectableNodes: [Array.from(this.nativeElement.childNodes)]
            });
            this.contentWrapperElement = this.contentWrapperRef.location.nativeElement;
            this.renderer.appendChild(this.nativeElement, this.contentWrapperElement);
        }
        this.appRef.attachView(this.contentWrapperRef.hostView);
    }
    attachScrollbars(hostElement) {
        // Create the scrollbars component
        this.scrollbarsRef = createComponent(Scrollbars, {
            elementInjector: this.injector,
            environmentInjector: this.appRef.injector
        });
        this.renderer.appendChild(hostElement, this.scrollbarsRef.location.nativeElement);
        // Attach the host view of the component to the main change detection tree, so that its lifecycle hooks run.
        this.appRef.attachView(this.scrollbarsRef.hostView);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollViewport, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.1.2", type: ScrollViewport, isStandalone: true, selector: "ng-scroll-viewport", inputs: { contentWrapperElement: "contentWrapperElement", spacerElement: "spacerElement" }, usesInheritance: true, ngImport: i0, template: '<ng-content/>', isInline: true, styles: [":host{--_scrollbar-track-thickness-px: var(--scrollbar-track-thickness, 7px);--_scrollbar-container-offset-px: var(--scrollbar-container-offset, 0px);--_scrollbar-total-thickness-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 2);--_scrollbar-thumb-min-size: var(--scrollbar-thumb-min-size, 20);position:relative!important;box-sizing:border-box!important;padding:0!important;-webkit-overflow-scrolling:touch;will-change:scroll-position}:host ::ng-deep>.ng-scroll-spacer{position:absolute;inset-block-start:0;inset-inline-start:0;width:calc(var(--_spacer-width) * 1px);height:calc(var(--_spacer-height) * 1px);min-width:100%;min-height:100%}:host{scrollbar-width:none!important}:host::-webkit-scrollbar{display:none!important}:host{--_viewport-overflow: auto;--_scrollbar-content-width: fit-content;overflow:var(--_viewport-overflow)}:host[orientation=vertical]{--_viewport-overflow: hidden auto;--_scrollbar-content-width: unset}:host[orientation=horizontal]{--_viewport-overflow: auto hidden}:host{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px);--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px);--_scrollbar-grid-template: \". scrollbar-y\" 1fr \"scrollbar-x .\" auto / 1fr auto;--_scrollbar-wrapper-y-start: initial;--_scrollbar-wrapper-y-end: 0;--_scrollbar-wrapper-x-top: initial;--_scrollbar-wrapper-x-bottom: 0}:host[position=invertY]{--_scrollbar-grid-template: \"scrollbar-y .\" 1fr \". scrollbar-x\" auto / auto 1fr}:host[position=invertX]{--_scrollbar-grid-template: \"scrollbar-x .\" auto \". scrollbar-y\" 1fr / 1fr auto}:host[position=invertAll]{--_scrollbar-grid-template: \". scrollbar-x\" auto \"scrollbar-y .\" 1fr / auto 1fr }:host{--_horizontal-block-start: initial;--_horizontal-block-end: var(--_scrollbar-container-offset-px)}:host:is([position=invertX],[position=invertAll]){--_horizontal-block-start: var(--_scrollbar-container-offset-px);--_horizontal-block-end: initial;--_scrollbar-wrapper-x-top: 0;--_scrollbar-wrapper-x-bottom: initial}:host{--_vertical-inset-inline-end: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-start: initial}:host:is([position=invertY],[position=invertAll]){--_vertical-inset-inline-start: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-end: initial;--_scrollbar-wrapper-y-start: 0;--_scrollbar-wrapper-y-end: initial}:host[verticalUsed=true][horizontalUsed=true]{--_scrollbar-thickness-margin-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 3)}:host[horizontalUsed=true]{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-thickness-margin-px)}:host[horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_vertical-inset-block-start: var(--_scrollbar-thickness-margin-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px)}:host[verticalUsed=true]{--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-thickness-margin-px)}:host[verticalUsed=true]:is([position=invertY],[position=invertAll]){--_horizontal-inline-start: var(--_scrollbar-thickness-margin-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px)}:host{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: 0;--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: 0}:host[appearance=native][verticalUsed=true]{--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][verticalUsed=true]:is([position=invertY],[position=invertAll]){--_viewport-padding-inline-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-inline-end: 0}:host[appearance=native][horizontalUsed=true]{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_viewport-padding-block-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-block-end: 0}:host{--_horizontal-thumb-display: block;--_vertical-thumb-display: block}:host[isVerticallyScrollable=false]{--_vertical-thumb-display: none}:host[isHorizontallyScrollable=false]{--_horizontal-thumb-display: none}:host[visibility=hover]{--_scrollbar-hover-opacity: 0;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-leave-duration, .4s) var(--scrollbar-hover-opacity-transition-leave-delay, 1s)}:host[visibility=hover][scrolling=true],:host[visibility=hover]:hover,:host[visibility=hover]:active,:host[visibility=hover]:focus{--_scrollbar-hover-opacity: 1;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-enter-duration, 0)}:host{--_viewport_scroll-timeline: unset;--_animation-timeline-y: unset;--_scrollbar-thumb-transform-from: unset;--_scrollbar-thumb-transform-to: unset}:host[verticalUsed=true]{--_timeline-scope: --scrollerY;--_animation-timeline-y: --scrollerY;--_viewport_scroll-timeline: --scrollerY y}:host[horizontalUsed=true]{--_timeline-scope: --scrollerX;--_animation-timeline-x: --scrollerX;--_viewport_scroll-timeline: --scrollerX x}:host[verticalUsed=true][horizontalUsed=true]{--_timeline-scope: --scrollerX, --scrollerY;--_viewport_scroll-timeline: --scrollerX x, --scrollerY y}:host{scroll-timeline:var(--_viewport_scroll-timeline);--_scrollbar-hover-thickness-px: var(--scrollbar-track-hover-thickness, var(--_scrollbar-track-thickness-px));--_scrollbar-active-thickness-px: var(--scrollbar-track-active-thickness, var(--_scrollbar-hover-thickness-px));--_track-y-thickness-px: var(--_scrollbar-track-thickness-px);--_track-x-thickness-px: var(--_scrollbar-track-thickness-px);-webkit-user-select:var(--_viewport-user-select);user-select:var(--_viewport-user-select)}:host[dragging=x],:host[dragging=y]{--_viewport-user-select: none}:host[dragging=x]{--_track-x-thickness-px: var(--_scrollbar-active-thickness-px)}:host[dragging=y]{--_track-y-thickness-px: var(--_scrollbar-active-thickness-px)}:host{overscroll-behavior:var(--scrollbar-overscroll-behavior, initial)}:host[mobile=true]{overscroll-behavior:var(--scrollbar-mobile-overscroll-behavior, none)}:host{--_viewport-pointer-events: auto;--_scrollbar-y-pointer-events: auto;--_scrollbar-x-pointer-events: auto;pointer-events:var(--_viewport-pointer-events)}:host[disableInteraction=true]{--_viewport-pointer-events: none;--_scrollbar-y-pointer-events: none;--_scrollbar-x-pointer-events: none}:host[isVerticallyScrollable=false]{--_scrollbar-y-pointer-events: none}:host[isHorizontallyScrollable=false]{--_scrollbar-x-pointer-events: none}:host{--_scrollbar-thumb-color: var(--scrollbar-thumb-color, #1e90ff);--_scrollbar-thumb-hover-color: var(--scrollbar-thumb-hover-color, var(--_scrollbar-thumb-color));--_scrollbar-thumb-active-color: var(--scrollbar-thumb-active-color, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-color: var(--scrollbar-button-color, transparent);--_scrollbar-button-hover-color: var(--scrollbar-button-hover-color, var(--_scrollbar-button-color));--_scrollbar-button-active-color: var(--scrollbar-button-active-color, var(--_scrollbar-button-hover-color));--_scrollbar-button-inactive-color: var(--scrollbar-button-inactive-color, color-mix(in srgb, var(--_scrollbar-button-color), #000 35%));--_scrollbar-button-fill: var(--scrollbar-button-fill, var(--_scrollbar-thumb-color));--_scrollbar-button-hover-fill: var(--scrollbar-button-hover-fill, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-active-fill: var(--scrollbar-button-active-fill, var(--_scrollbar-thumb-active-color));--_scrollbar-button-inactive-fill: var(--scrollbar-button-inactive-fill, color-mix(in srgb, var(--_scrollbar-button-fill), #000 35%))}@layer default{:host{display:block;max-height:100%;max-width:100%}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: ScrollViewport, decorators: [{
            type: Component,
            args: [{ selector: 'ng-scroll-viewport', template: '<ng-content/>', changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{--_scrollbar-track-thickness-px: var(--scrollbar-track-thickness, 7px);--_scrollbar-container-offset-px: var(--scrollbar-container-offset, 0px);--_scrollbar-total-thickness-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 2);--_scrollbar-thumb-min-size: var(--scrollbar-thumb-min-size, 20);position:relative!important;box-sizing:border-box!important;padding:0!important;-webkit-overflow-scrolling:touch;will-change:scroll-position}:host ::ng-deep>.ng-scroll-spacer{position:absolute;inset-block-start:0;inset-inline-start:0;width:calc(var(--_spacer-width) * 1px);height:calc(var(--_spacer-height) * 1px);min-width:100%;min-height:100%}:host{scrollbar-width:none!important}:host::-webkit-scrollbar{display:none!important}:host{--_viewport-overflow: auto;--_scrollbar-content-width: fit-content;overflow:var(--_viewport-overflow)}:host[orientation=vertical]{--_viewport-overflow: hidden auto;--_scrollbar-content-width: unset}:host[orientation=horizontal]{--_viewport-overflow: auto hidden}:host{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px);--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px);--_scrollbar-grid-template: \". scrollbar-y\" 1fr \"scrollbar-x .\" auto / 1fr auto;--_scrollbar-wrapper-y-start: initial;--_scrollbar-wrapper-y-end: 0;--_scrollbar-wrapper-x-top: initial;--_scrollbar-wrapper-x-bottom: 0}:host[position=invertY]{--_scrollbar-grid-template: \"scrollbar-y .\" 1fr \". scrollbar-x\" auto / auto 1fr}:host[position=invertX]{--_scrollbar-grid-template: \"scrollbar-x .\" auto \". scrollbar-y\" 1fr / 1fr auto}:host[position=invertAll]{--_scrollbar-grid-template: \". scrollbar-x\" auto \"scrollbar-y .\" 1fr / auto 1fr }:host{--_horizontal-block-start: initial;--_horizontal-block-end: var(--_scrollbar-container-offset-px)}:host:is([position=invertX],[position=invertAll]){--_horizontal-block-start: var(--_scrollbar-container-offset-px);--_horizontal-block-end: initial;--_scrollbar-wrapper-x-top: 0;--_scrollbar-wrapper-x-bottom: initial}:host{--_vertical-inset-inline-end: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-start: initial}:host:is([position=invertY],[position=invertAll]){--_vertical-inset-inline-start: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-end: initial;--_scrollbar-wrapper-y-start: 0;--_scrollbar-wrapper-y-end: initial}:host[verticalUsed=true][horizontalUsed=true]{--_scrollbar-thickness-margin-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 3)}:host[horizontalUsed=true]{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-thickness-margin-px)}:host[horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_vertical-inset-block-start: var(--_scrollbar-thickness-margin-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px)}:host[verticalUsed=true]{--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-thickness-margin-px)}:host[verticalUsed=true]:is([position=invertY],[position=invertAll]){--_horizontal-inline-start: var(--_scrollbar-thickness-margin-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px)}:host{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: 0;--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: 0}:host[appearance=native][verticalUsed=true]{--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][verticalUsed=true]:is([position=invertY],[position=invertAll]){--_viewport-padding-inline-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-inline-end: 0}:host[appearance=native][horizontalUsed=true]{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_viewport-padding-block-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-block-end: 0}:host{--_horizontal-thumb-display: block;--_vertical-thumb-display: block}:host[isVerticallyScrollable=false]{--_vertical-thumb-display: none}:host[isHorizontallyScrollable=false]{--_horizontal-thumb-display: none}:host[visibility=hover]{--_scrollbar-hover-opacity: 0;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-leave-duration, .4s) var(--scrollbar-hover-opacity-transition-leave-delay, 1s)}:host[visibility=hover][scrolling=true],:host[visibility=hover]:hover,:host[visibility=hover]:active,:host[visibility=hover]:focus{--_scrollbar-hover-opacity: 1;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-enter-duration, 0)}:host{--_viewport_scroll-timeline: unset;--_animation-timeline-y: unset;--_scrollbar-thumb-transform-from: unset;--_scrollbar-thumb-transform-to: unset}:host[verticalUsed=true]{--_timeline-scope: --scrollerY;--_animation-timeline-y: --scrollerY;--_viewport_scroll-timeline: --scrollerY y}:host[horizontalUsed=true]{--_timeline-scope: --scrollerX;--_animation-timeline-x: --scrollerX;--_viewport_scroll-timeline: --scrollerX x}:host[verticalUsed=true][horizontalUsed=true]{--_timeline-scope: --scrollerX, --scrollerY;--_viewport_scroll-timeline: --scrollerX x, --scrollerY y}:host{scroll-timeline:var(--_viewport_scroll-timeline);--_scrollbar-hover-thickness-px: var(--scrollbar-track-hover-thickness, var(--_scrollbar-track-thickness-px));--_scrollbar-active-thickness-px: var(--scrollbar-track-active-thickness, var(--_scrollbar-hover-thickness-px));--_track-y-thickness-px: var(--_scrollbar-track-thickness-px);--_track-x-thickness-px: var(--_scrollbar-track-thickness-px);-webkit-user-select:var(--_viewport-user-select);user-select:var(--_viewport-user-select)}:host[dragging=x],:host[dragging=y]{--_viewport-user-select: none}:host[dragging=x]{--_track-x-thickness-px: var(--_scrollbar-active-thickness-px)}:host[dragging=y]{--_track-y-thickness-px: var(--_scrollbar-active-thickness-px)}:host{overscroll-behavior:var(--scrollbar-overscroll-behavior, initial)}:host[mobile=true]{overscroll-behavior:var(--scrollbar-mobile-overscroll-behavior, none)}:host{--_viewport-pointer-events: auto;--_scrollbar-y-pointer-events: auto;--_scrollbar-x-pointer-events: auto;pointer-events:var(--_viewport-pointer-events)}:host[disableInteraction=true]{--_viewport-pointer-events: none;--_scrollbar-y-pointer-events: none;--_scrollbar-x-pointer-events: none}:host[isVerticallyScrollable=false]{--_scrollbar-y-pointer-events: none}:host[isHorizontallyScrollable=false]{--_scrollbar-x-pointer-events: none}:host{--_scrollbar-thumb-color: var(--scrollbar-thumb-color, #1e90ff);--_scrollbar-thumb-hover-color: var(--scrollbar-thumb-hover-color, var(--_scrollbar-thumb-color));--_scrollbar-thumb-active-color: var(--scrollbar-thumb-active-color, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-color: var(--scrollbar-button-color, transparent);--_scrollbar-button-hover-color: var(--scrollbar-button-hover-color, var(--_scrollbar-button-color));--_scrollbar-button-active-color: var(--scrollbar-button-active-color, var(--_scrollbar-button-hover-color));--_scrollbar-button-inactive-color: var(--scrollbar-button-inactive-color, color-mix(in srgb, var(--_scrollbar-button-color), #000 35%));--_scrollbar-button-fill: var(--scrollbar-button-fill, var(--_scrollbar-thumb-color));--_scrollbar-button-hover-fill: var(--scrollbar-button-hover-fill, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-active-fill: var(--scrollbar-button-active-fill, var(--_scrollbar-thumb-active-color));--_scrollbar-button-inactive-fill: var(--scrollbar-button-inactive-fill, color-mix(in srgb, var(--_scrollbar-button-fill), #000 35%))}@layer default{:host{display:block;max-height:100%;max-width:100%}}\n"] }]
        }], ctorParameters: () => [], propDecorators: { contentWrapperElement: [{
                type: Input
            }], spacerElement: [{
                type: Input
            }] } });

class NgScrollbar extends NgScrollbarCore {
    contentWrapper = viewChild.required(ScrollContent, { read: ElementRef });
    contentWrapperElement = computed(() => this.contentWrapper().nativeElement, ...(ngDevMode ? [{ debugName: "contentWrapperElement" }] : []));
    constructor() {
        effect(() => {
            const contentWrapper = this.contentWrapperElement();
            untracked(() => {
                this.adapter.init(this.nativeElement, contentWrapper);
            });
        });
        super();
    }
    /**
     * Smooth scroll functions
     */
    scrollTo(options) {
        return this.adapter.scrollTo(options);
    }
    /**
     * Scroll to an element by reference or selector
     */
    scrollToElement(target, options) {
        return this.adapter.scrollToElement(target, options);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbar, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "21.1.2", type: NgScrollbar, isStandalone: true, selector: "ng-scrollbar:not([externalViewport]), [ngScrollbar]", host: { properties: { "class.ng-scrollbar": "true" } }, viewQueries: [{ propertyName: "contentWrapper", first: true, predicate: ScrollContent, descendants: true, read: ElementRef, isSignal: true }], exportAs: ["ngScrollbar"], usesInheritance: true, hostDirectives: [{ directive: ViewportAdapter, inputs: ["position", "position", "withButtons", "withButtons", "appearance", "appearance", "thumbClass", "thumbClass", "trackClass", "trackClass", "buttonClass", "buttonClass", "visibility", "visibility", "hoverOffset", "hoverOffset", "orientation", "orientation", "disableSensor", "disableSensor", "sensorThrottleTime", "sensorThrottleTime", "disableInteraction", "disableInteraction", "trackScrollDuration", "trackScrollDuration"], outputs: ["afterInit", "afterInit", "afterUpdate", "afterUpdate"] }], ngImport: i0, template: `
    <ng-scroll-content>
      <ng-content/>
      <scrollbars/>
    </ng-scroll-content>
  `, isInline: true, styles: [":host{--_scrollbar-track-thickness-px: var(--scrollbar-track-thickness, 7px);--_scrollbar-container-offset-px: var(--scrollbar-container-offset, 0px);--_scrollbar-total-thickness-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 2);--_scrollbar-thumb-min-size: var(--scrollbar-thumb-min-size, 20);position:relative!important;box-sizing:border-box!important;padding:0!important;-webkit-overflow-scrolling:touch;will-change:scroll-position}:host ::ng-deep>.ng-scroll-spacer{position:absolute;inset-block-start:0;inset-inline-start:0;width:calc(var(--_spacer-width) * 1px);height:calc(var(--_spacer-height) * 1px);min-width:100%;min-height:100%}:host{scrollbar-width:none!important}:host::-webkit-scrollbar{display:none!important}:host{--_viewport-overflow: auto;--_scrollbar-content-width: fit-content;overflow:var(--_viewport-overflow)}:host[orientation=vertical]{--_viewport-overflow: hidden auto;--_scrollbar-content-width: unset}:host[orientation=horizontal]{--_viewport-overflow: auto hidden}:host{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px);--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px);--_scrollbar-grid-template: \". scrollbar-y\" 1fr \"scrollbar-x .\" auto / 1fr auto;--_scrollbar-wrapper-y-start: initial;--_scrollbar-wrapper-y-end: 0;--_scrollbar-wrapper-x-top: initial;--_scrollbar-wrapper-x-bottom: 0}:host[position=invertY]{--_scrollbar-grid-template: \"scrollbar-y .\" 1fr \". scrollbar-x\" auto / auto 1fr}:host[position=invertX]{--_scrollbar-grid-template: \"scrollbar-x .\" auto \". scrollbar-y\" 1fr / 1fr auto}:host[position=invertAll]{--_scrollbar-grid-template: \". scrollbar-x\" auto \"scrollbar-y .\" 1fr / auto 1fr }:host{--_horizontal-block-start: initial;--_horizontal-block-end: var(--_scrollbar-container-offset-px)}:host:is([position=invertX],[position=invertAll]){--_horizontal-block-start: var(--_scrollbar-container-offset-px);--_horizontal-block-end: initial;--_scrollbar-wrapper-x-top: 0;--_scrollbar-wrapper-x-bottom: initial}:host{--_vertical-inset-inline-end: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-start: initial}:host:is([position=invertY],[position=invertAll]){--_vertical-inset-inline-start: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-end: initial;--_scrollbar-wrapper-y-start: 0;--_scrollbar-wrapper-y-end: initial}:host[verticalUsed=true][horizontalUsed=true]{--_scrollbar-thickness-margin-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 3)}:host[horizontalUsed=true]{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-thickness-margin-px)}:host[horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_vertical-inset-block-start: var(--_scrollbar-thickness-margin-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px)}:host[verticalUsed=true]{--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-thickness-margin-px)}:host[verticalUsed=true]:is([position=invertY],[position=invertAll]){--_horizontal-inline-start: var(--_scrollbar-thickness-margin-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px)}:host{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: 0;--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: 0}:host[appearance=native][verticalUsed=true]{--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][verticalUsed=true]:is([position=invertY],[position=invertAll]){--_viewport-padding-inline-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-inline-end: 0}:host[appearance=native][horizontalUsed=true]{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_viewport-padding-block-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-block-end: 0}:host{--_horizontal-thumb-display: block;--_vertical-thumb-display: block}:host[isVerticallyScrollable=false]{--_vertical-thumb-display: none}:host[isHorizontallyScrollable=false]{--_horizontal-thumb-display: none}:host[visibility=hover]{--_scrollbar-hover-opacity: 0;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-leave-duration, .4s) var(--scrollbar-hover-opacity-transition-leave-delay, 1s)}:host[visibility=hover][scrolling=true],:host[visibility=hover]:hover,:host[visibility=hover]:active,:host[visibility=hover]:focus{--_scrollbar-hover-opacity: 1;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-enter-duration, 0)}:host{--_viewport_scroll-timeline: unset;--_animation-timeline-y: unset;--_scrollbar-thumb-transform-from: unset;--_scrollbar-thumb-transform-to: unset}:host[verticalUsed=true]{--_timeline-scope: --scrollerY;--_animation-timeline-y: --scrollerY;--_viewport_scroll-timeline: --scrollerY y}:host[horizontalUsed=true]{--_timeline-scope: --scrollerX;--_animation-timeline-x: --scrollerX;--_viewport_scroll-timeline: --scrollerX x}:host[verticalUsed=true][horizontalUsed=true]{--_timeline-scope: --scrollerX, --scrollerY;--_viewport_scroll-timeline: --scrollerX x, --scrollerY y}:host{scroll-timeline:var(--_viewport_scroll-timeline);--_scrollbar-hover-thickness-px: var(--scrollbar-track-hover-thickness, var(--_scrollbar-track-thickness-px));--_scrollbar-active-thickness-px: var(--scrollbar-track-active-thickness, var(--_scrollbar-hover-thickness-px));--_track-y-thickness-px: var(--_scrollbar-track-thickness-px);--_track-x-thickness-px: var(--_scrollbar-track-thickness-px);-webkit-user-select:var(--_viewport-user-select);user-select:var(--_viewport-user-select)}:host[dragging=x],:host[dragging=y]{--_viewport-user-select: none}:host[dragging=x]{--_track-x-thickness-px: var(--_scrollbar-active-thickness-px)}:host[dragging=y]{--_track-y-thickness-px: var(--_scrollbar-active-thickness-px)}:host{overscroll-behavior:var(--scrollbar-overscroll-behavior, initial)}:host[mobile=true]{overscroll-behavior:var(--scrollbar-mobile-overscroll-behavior, none)}:host{--_viewport-pointer-events: auto;--_scrollbar-y-pointer-events: auto;--_scrollbar-x-pointer-events: auto;pointer-events:var(--_viewport-pointer-events)}:host[disableInteraction=true]{--_viewport-pointer-events: none;--_scrollbar-y-pointer-events: none;--_scrollbar-x-pointer-events: none}:host[isVerticallyScrollable=false]{--_scrollbar-y-pointer-events: none}:host[isHorizontallyScrollable=false]{--_scrollbar-x-pointer-events: none}:host{--_scrollbar-thumb-color: var(--scrollbar-thumb-color, #1e90ff);--_scrollbar-thumb-hover-color: var(--scrollbar-thumb-hover-color, var(--_scrollbar-thumb-color));--_scrollbar-thumb-active-color: var(--scrollbar-thumb-active-color, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-color: var(--scrollbar-button-color, transparent);--_scrollbar-button-hover-color: var(--scrollbar-button-hover-color, var(--_scrollbar-button-color));--_scrollbar-button-active-color: var(--scrollbar-button-active-color, var(--_scrollbar-button-hover-color));--_scrollbar-button-inactive-color: var(--scrollbar-button-inactive-color, color-mix(in srgb, var(--_scrollbar-button-color), #000 35%));--_scrollbar-button-fill: var(--scrollbar-button-fill, var(--_scrollbar-thumb-color));--_scrollbar-button-hover-fill: var(--scrollbar-button-hover-fill, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-active-fill: var(--scrollbar-button-active-fill, var(--_scrollbar-thumb-active-color));--_scrollbar-button-inactive-fill: var(--scrollbar-button-inactive-fill, color-mix(in srgb, var(--_scrollbar-button-fill), #000 35%))}@layer default{:host{display:block;max-height:100%;max-width:100%}}\n"], dependencies: [{ kind: "component", type: Scrollbars, selector: "scrollbars" }, { kind: "component", type: ScrollContent, selector: "ng-scroll-content" }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbar, decorators: [{
            type: Component,
            args: [{ host: {
                        '[class.ng-scrollbar]': 'true'
                    }, selector: 'ng-scrollbar:not([externalViewport]), [ngScrollbar]', exportAs: 'ngScrollbar', imports: [Scrollbars, ScrollContent], template: `
    <ng-scroll-content>
      <ng-content/>
      <scrollbars/>
    </ng-scroll-content>
  `, changeDetection: ChangeDetectionStrategy.OnPush, hostDirectives: [{
                            directive: ViewportAdapter,
                            inputs: [
                                'position',
                                'withButtons',
                                'appearance',
                                'thumbClass',
                                'trackClass',
                                'buttonClass',
                                'visibility',
                                'hoverOffset',
                                'orientation',
                                'disableSensor',
                                'sensorThrottleTime',
                                'disableInteraction',
                                'trackScrollDuration'
                            ],
                            outputs: ['afterInit', 'afterUpdate']
                        }], styles: [":host{--_scrollbar-track-thickness-px: var(--scrollbar-track-thickness, 7px);--_scrollbar-container-offset-px: var(--scrollbar-container-offset, 0px);--_scrollbar-total-thickness-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 2);--_scrollbar-thumb-min-size: var(--scrollbar-thumb-min-size, 20);position:relative!important;box-sizing:border-box!important;padding:0!important;-webkit-overflow-scrolling:touch;will-change:scroll-position}:host ::ng-deep>.ng-scroll-spacer{position:absolute;inset-block-start:0;inset-inline-start:0;width:calc(var(--_spacer-width) * 1px);height:calc(var(--_spacer-height) * 1px);min-width:100%;min-height:100%}:host{scrollbar-width:none!important}:host::-webkit-scrollbar{display:none!important}:host{--_viewport-overflow: auto;--_scrollbar-content-width: fit-content;overflow:var(--_viewport-overflow)}:host[orientation=vertical]{--_viewport-overflow: hidden auto;--_scrollbar-content-width: unset}:host[orientation=horizontal]{--_viewport-overflow: auto hidden}:host{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px);--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px);--_scrollbar-grid-template: \". scrollbar-y\" 1fr \"scrollbar-x .\" auto / 1fr auto;--_scrollbar-wrapper-y-start: initial;--_scrollbar-wrapper-y-end: 0;--_scrollbar-wrapper-x-top: initial;--_scrollbar-wrapper-x-bottom: 0}:host[position=invertY]{--_scrollbar-grid-template: \"scrollbar-y .\" 1fr \". scrollbar-x\" auto / auto 1fr}:host[position=invertX]{--_scrollbar-grid-template: \"scrollbar-x .\" auto \". scrollbar-y\" 1fr / 1fr auto}:host[position=invertAll]{--_scrollbar-grid-template: \". scrollbar-x\" auto \"scrollbar-y .\" 1fr / auto 1fr }:host{--_horizontal-block-start: initial;--_horizontal-block-end: var(--_scrollbar-container-offset-px)}:host:is([position=invertX],[position=invertAll]){--_horizontal-block-start: var(--_scrollbar-container-offset-px);--_horizontal-block-end: initial;--_scrollbar-wrapper-x-top: 0;--_scrollbar-wrapper-x-bottom: initial}:host{--_vertical-inset-inline-end: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-start: initial}:host:is([position=invertY],[position=invertAll]){--_vertical-inset-inline-start: var(--_scrollbar-container-offset-px);--_vertical-inset-inline-end: initial;--_scrollbar-wrapper-y-start: 0;--_scrollbar-wrapper-y-end: initial}:host[verticalUsed=true][horizontalUsed=true]{--_scrollbar-thickness-margin-px: calc(var(--_scrollbar-track-thickness-px) + var(--_scrollbar-container-offset-px) * 3)}:host[horizontalUsed=true]{--_vertical-inset-block-start: var(--_scrollbar-container-offset-px);--_vertical-inset-block-end: var(--_scrollbar-thickness-margin-px)}:host[horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_vertical-inset-block-start: var(--_scrollbar-thickness-margin-px);--_vertical-inset-block-end: var(--_scrollbar-container-offset-px)}:host[verticalUsed=true]{--_horizontal-inline-start: var(--_scrollbar-container-offset-px);--_horizontal-inline-end: var(--_scrollbar-thickness-margin-px)}:host[verticalUsed=true]:is([position=invertY],[position=invertAll]){--_horizontal-inline-start: var(--_scrollbar-thickness-margin-px);--_horizontal-inline-end: var(--_scrollbar-container-offset-px)}:host{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: 0;--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: 0}:host[appearance=native][verticalUsed=true]{--_viewport-padding-inline-start: 0;--_viewport-padding-inline-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][verticalUsed=true]:is([position=invertY],[position=invertAll]){--_viewport-padding-inline-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-inline-end: 0}:host[appearance=native][horizontalUsed=true]{--_viewport-padding-block-start: 0;--_viewport-padding-block-end: var(--_scrollbar-total-thickness-px)}:host[appearance=native][horizontalUsed=true]:is([position=invertX],[position=invertAll]){--_viewport-padding-block-start: var(--_scrollbar-total-thickness-px);--_viewport-padding-block-end: 0}:host{--_horizontal-thumb-display: block;--_vertical-thumb-display: block}:host[isVerticallyScrollable=false]{--_vertical-thumb-display: none}:host[isHorizontallyScrollable=false]{--_horizontal-thumb-display: none}:host[visibility=hover]{--_scrollbar-hover-opacity: 0;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-leave-duration, .4s) var(--scrollbar-hover-opacity-transition-leave-delay, 1s)}:host[visibility=hover][scrolling=true],:host[visibility=hover]:hover,:host[visibility=hover]:active,:host[visibility=hover]:focus{--_scrollbar-hover-opacity: 1;--_scrollbar-opacity-transition: opacity var(--scrollbar-hover-opacity-transition-enter-duration, 0)}:host{--_viewport_scroll-timeline: unset;--_animation-timeline-y: unset;--_scrollbar-thumb-transform-from: unset;--_scrollbar-thumb-transform-to: unset}:host[verticalUsed=true]{--_timeline-scope: --scrollerY;--_animation-timeline-y: --scrollerY;--_viewport_scroll-timeline: --scrollerY y}:host[horizontalUsed=true]{--_timeline-scope: --scrollerX;--_animation-timeline-x: --scrollerX;--_viewport_scroll-timeline: --scrollerX x}:host[verticalUsed=true][horizontalUsed=true]{--_timeline-scope: --scrollerX, --scrollerY;--_viewport_scroll-timeline: --scrollerX x, --scrollerY y}:host{scroll-timeline:var(--_viewport_scroll-timeline);--_scrollbar-hover-thickness-px: var(--scrollbar-track-hover-thickness, var(--_scrollbar-track-thickness-px));--_scrollbar-active-thickness-px: var(--scrollbar-track-active-thickness, var(--_scrollbar-hover-thickness-px));--_track-y-thickness-px: var(--_scrollbar-track-thickness-px);--_track-x-thickness-px: var(--_scrollbar-track-thickness-px);-webkit-user-select:var(--_viewport-user-select);user-select:var(--_viewport-user-select)}:host[dragging=x],:host[dragging=y]{--_viewport-user-select: none}:host[dragging=x]{--_track-x-thickness-px: var(--_scrollbar-active-thickness-px)}:host[dragging=y]{--_track-y-thickness-px: var(--_scrollbar-active-thickness-px)}:host{overscroll-behavior:var(--scrollbar-overscroll-behavior, initial)}:host[mobile=true]{overscroll-behavior:var(--scrollbar-mobile-overscroll-behavior, none)}:host{--_viewport-pointer-events: auto;--_scrollbar-y-pointer-events: auto;--_scrollbar-x-pointer-events: auto;pointer-events:var(--_viewport-pointer-events)}:host[disableInteraction=true]{--_viewport-pointer-events: none;--_scrollbar-y-pointer-events: none;--_scrollbar-x-pointer-events: none}:host[isVerticallyScrollable=false]{--_scrollbar-y-pointer-events: none}:host[isHorizontallyScrollable=false]{--_scrollbar-x-pointer-events: none}:host{--_scrollbar-thumb-color: var(--scrollbar-thumb-color, #1e90ff);--_scrollbar-thumb-hover-color: var(--scrollbar-thumb-hover-color, var(--_scrollbar-thumb-color));--_scrollbar-thumb-active-color: var(--scrollbar-thumb-active-color, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-color: var(--scrollbar-button-color, transparent);--_scrollbar-button-hover-color: var(--scrollbar-button-hover-color, var(--_scrollbar-button-color));--_scrollbar-button-active-color: var(--scrollbar-button-active-color, var(--_scrollbar-button-hover-color));--_scrollbar-button-inactive-color: var(--scrollbar-button-inactive-color, color-mix(in srgb, var(--_scrollbar-button-color), #000 35%));--_scrollbar-button-fill: var(--scrollbar-button-fill, var(--_scrollbar-thumb-color));--_scrollbar-button-hover-fill: var(--scrollbar-button-hover-fill, var(--_scrollbar-thumb-hover-color));--_scrollbar-button-active-fill: var(--scrollbar-button-active-fill, var(--_scrollbar-thumb-active-color));--_scrollbar-button-inactive-fill: var(--scrollbar-button-inactive-fill, color-mix(in srgb, var(--_scrollbar-button-fill), #000 35%))}@layer default{:host{display:block;max-height:100%;max-width:100%}}\n"] }]
        }], ctorParameters: () => [], propDecorators: { contentWrapper: [{ type: i0.ViewChild, args: [i0.forwardRef(() => ScrollContent), { ...{ read: ElementRef }, isSignal: true }] }] } });

class NgScrollbarExt {
    appRef = inject(ApplicationRef);
    injector = inject(Injector);
    adapter = inject(ViewportAdapter);
    nativeElement = inject(ElementRef).nativeElement;
    viewportRef;
    /**
     * Selector used to query the viewport element.
     */
    externalViewport = input.required(...(ngDevMode ? [{ debugName: "externalViewport" }] : []));
    /**
     * Selector used to query the content wrapper element.
     */
    externalContentWrapper = input(...(ngDevMode ? [undefined, { debugName: "externalContentWrapper" }] : []));
    /**
     * Selector used to query the spacer element (virtual scroll integration).
     * In the case of integrating the scrollbar with a virtual scroll component,
     * a spacer element is typically created to match the real size of the content.
     * The scrollbar will use the size of this spacer element for calculations instead of the content wrapper size.
     */
    externalSpacer = input(...(ngDevMode ? [undefined, { debugName: "externalSpacer" }] : []));
    viewportElement = linkedSignal({ ...(ngDevMode ? { debugName: "viewportElement" } : {}), source: this.externalViewport,
        // If viewport selector was defined, query the element
        computation: (selector) => this._getElement(selector) });
    viewportError = computed(() => {
        return !this.viewportElement()
            ? `[NgScrollbar]: Could not find the viewport element for the provided selector "${this.externalViewport()}"`
            : null;
    }, ...(ngDevMode ? [{ debugName: "viewportError" }] : []));
    contentWrapperElement = linkedSignal({ ...(ngDevMode ? { debugName: "contentWrapperElement" } : {}), source: this.externalContentWrapper,
        computation: (selector) => this._getElement(selector) });
    contentWrapperError = computed(() => {
        return !this.contentWrapperElement() && this.externalContentWrapper()
            ? `[NgScrollbar]: Content wrapper element not found for the provided selector "${this.externalContentWrapper()}"`
            : null;
    }, ...(ngDevMode ? [{ debugName: "contentWrapperError" }] : []));
    spacerElement = linkedSignal({ ...(ngDevMode ? { debugName: "spacerElement" } : {}), source: this.externalSpacer,
        computation: (selector) => this._getElement(selector) });
    spacerError = computed(() => {
        return !this.spacerElement() && this.externalSpacer()
            ? `[NgScrollbar]: Spacer element not found for the provided selector "${this.externalSpacer()}"`
            : null;
    }, ...(ngDevMode ? [{ debugName: "spacerError" }] : []));
    /**
     * Skip initializing the viewport and the scrollbar
     * this is used when the component needs to wait for 3rd party library to render
     */
    skipInit = false;
    constructor() {
        afterNextRender({
            earlyRead: () => {
                if (this.skipInit)
                    return;
                const viewportElement = this.viewportElement();
                const contentWrapperElement = this.contentWrapperElement();
                const spacerElement = this.spacerElement();
                const viewportError = this.viewportError();
                const contentWrapperError = this.contentWrapperError();
                const spacerError = this.spacerError();
                untracked(() => {
                    const error = viewportError || contentWrapperError || spacerError;
                    if (error) {
                        console.error(error);
                    }
                    else {
                        this.initialize(viewportElement, contentWrapperElement, spacerElement);
                    }
                });
            }
        });
    }
    ngOnDestroy() {
        this._destroy();
    }
    initialize(viewportElement, contentWrapperElement, spacerElement) {
        this.viewportRef = createComponent(ScrollViewport, {
            hostElement: viewportElement,
            projectableNodes: [Array.from(viewportElement.childNodes)],
            environmentInjector: this.appRef.injector,
            elementInjector: this.injector,
            bindings: [
                inputBinding('contentWrapperElement', () => contentWrapperElement),
                inputBinding('spacerElement', () => spacerElement)
            ]
        });
        this.appRef.attachView(this.viewportRef.hostView);
    }
    _destroy() {
        if (this.viewportRef) {
            this.appRef.detachView(this.viewportRef.hostView);
            this.viewportRef.destroy();
            this.viewportRef = null;
        }
    }
    _getElement(selector) {
        return selector ? this.nativeElement.querySelector(selector) : null;
    }
    /**
     * Smooth scroll functions
     */
    scrollTo(options) {
        return this.adapter.scrollTo(options);
    }
    /**
     * Scroll to an element by reference or selector
     */
    scrollToElement(target, options) {
        return this.adapter.scrollToElement(target, options);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarExt, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.1.2", type: NgScrollbarExt, isStandalone: true, selector: "ng-scrollbar[externalViewport]", inputs: { externalViewport: { classPropertyName: "externalViewport", publicName: "externalViewport", isSignal: true, isRequired: true, transformFunction: null }, externalContentWrapper: { classPropertyName: "externalContentWrapper", publicName: "externalContentWrapper", isSignal: true, isRequired: false, transformFunction: null }, externalSpacer: { classPropertyName: "externalSpacer", publicName: "externalSpacer", isSignal: true, isRequired: false, transformFunction: null } }, exportAs: ["ngScrollbar"], hostDirectives: [{ directive: ViewportAdapter, inputs: ["position", "position", "withButtons", "withButtons", "appearance", "appearance", "thumbClass", "thumbClass", "trackClass", "trackClass", "buttonClass", "buttonClass", "visibility", "visibility", "hoverOffset", "hoverOffset", "orientation", "orientation", "disableSensor", "disableSensor", "sensorThrottleTime", "sensorThrottleTime", "disableInteraction", "disableInteraction", "trackScrollDuration", "trackScrollDuration"], outputs: ["afterInit", "afterInit", "afterUpdate", "afterUpdate"] }], ngImport: i0, template: '<ng-content/>', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarExt, decorators: [{
            type: Component,
            args: [{
                    selector: 'ng-scrollbar[externalViewport]',
                    exportAs: 'ngScrollbar',
                    template: '<ng-content/>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    hostDirectives: [
                        {
                            directive: ViewportAdapter,
                            inputs: [
                                'position',
                                'withButtons',
                                'appearance',
                                'thumbClass',
                                'trackClass',
                                'buttonClass',
                                'visibility',
                                'hoverOffset',
                                'orientation',
                                'disableSensor',
                                'sensorThrottleTime',
                                'disableInteraction',
                                'trackScrollDuration'
                            ],
                            outputs: ['afterInit', 'afterUpdate']
                        }
                    ]
                }]
        }], ctorParameters: () => [], propDecorators: { externalViewport: [{ type: i0.Input, args: [{ isSignal: true, alias: "externalViewport", required: true }] }], externalContentWrapper: [{ type: i0.Input, args: [{ isSignal: true, alias: "externalContentWrapper", required: false }] }], externalSpacer: [{ type: i0.Input, args: [{ isSignal: true, alias: "externalSpacer", required: false }] }] } });

class NgScrollbarAsyncViewport {
    scrollbar = inject(NgScrollbarExt, { self: true });
    contentObserver = inject(ContentObserver);
    asyncViewport = input.required(...(ngDevMode ? [{ debugName: "asyncViewport" }] : []));
    constructor() {
        this.scrollbar.skipInit = true;
        let sub$;
        let init;
        effect((onCleanup) => {
            const externalViewport = this.scrollbar.externalViewport();
            const externalContentWrapper = this.scrollbar.externalContentWrapper();
            const externalSpacer = this.scrollbar.externalSpacer();
            const asyncDetection = this.asyncViewport();
            // The content observer should not be throttled using the same function we use for ResizeObserver,
            // It should detect the content change asap to attach the scrollbar
            sub$ = this.contentObserver.observe(this.scrollbar.nativeElement).pipe(throttleTime(100, null, {
                leading: true,
                trailing: true
            })).subscribe(() => {
                // Search for the viewport element
                this.scrollbar.viewportElement.set(this.scrollbar._getElement(externalViewport));
                // Search for the content wrapper element
                this.scrollbar.contentWrapperElement.set(this.scrollbar._getElement(externalContentWrapper));
                // Search for the spacer element
                this.scrollbar.spacerElement.set(this.scrollbar._getElement(externalSpacer));
                const contentWrapperCheck = externalContentWrapper ? !!this.scrollbar.contentWrapperElement() : true;
                const spacerPassCheck = externalSpacer ? !!this.scrollbar.spacerElement() : true;
                if (!init && this.scrollbar.viewportElement() && contentWrapperCheck && spacerPassCheck) {
                    // If an external spacer selector is provided, search for it
                    this.scrollbar.initialize(this.scrollbar.viewportElement(), this.scrollbar.contentWrapperElement(), this.scrollbar.spacerElement());
                    init = true;
                }
                else if (!this.scrollbar.viewportElement() ||
                    (externalContentWrapper && !this.scrollbar.contentWrapperElement()) ||
                    (externalSpacer && !this.scrollbar.spacerElement())) {
                    this.scrollbar._destroy();
                    init = false;
                }
                if (asyncDetection !== 'auto') {
                    sub$.unsubscribe();
                }
            });
            onCleanup(() => sub$?.unsubscribe());
        });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarAsyncViewport, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "21.1.2", type: NgScrollbarAsyncViewport, isStandalone: true, selector: "ng-scrollbar[externalViewport][asyncViewport]", inputs: { asyncViewport: { classPropertyName: "asyncViewport", publicName: "asyncViewport", isSignal: true, isRequired: true, transformFunction: null } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarAsyncViewport, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-scrollbar[externalViewport][asyncViewport]'
                }]
        }], ctorParameters: () => [], propDecorators: { asyncViewport: [{ type: i0.Input, args: [{ isSignal: true, alias: "asyncViewport", required: true }] }] } });

class NgSyncSpacer {
    zone = inject(NgZone);
    sharedResizeObserver = inject(SharedResizeObserver);
    scrollbar = inject(NgScrollbarExt, { self: true });
    viewport = inject(ViewportAdapter, { self: true });
    /**
     * A signal used to sync spacer dimension when content dimension changes
     */
    spacerDimension = signal({}, ...(ngDevMode ? [{ debugName: "spacerDimension" }] : []));
    constructor() {
        let resizeSub$;
        effect((onCleanup) => {
            const throttleDuration = this.viewport.sensorThrottleTime();
            const disableSensor = this.viewport.disableSensor();
            const contentWrapperElement = this.scrollbar.contentWrapperElement();
            const spacerElement = this.scrollbar.spacerElement();
            untracked(() => {
                if (!disableSensor && contentWrapperElement && spacerElement) {
                    // Sync spacer dimension with content wrapper dimensions to allow both scrollbars to be displayed
                    this.zone.runOutsideAngular(() => {
                        resizeSub$ = getThrottledStream(this.sharedResizeObserver.observe(contentWrapperElement), throttleDuration).subscribe(() => {
                            this.zone.run(() => {
                                // Use animation frame to avoid "ResizeObserver loop completed with undelivered notifications." error
                                requestAnimationFrame(() => {
                                    this.spacerDimension.set({
                                        width: contentWrapperElement.offsetWidth,
                                        height: contentWrapperElement.offsetHeight
                                    });
                                });
                            });
                        });
                    });
                }
                onCleanup(() => resizeSub$?.unsubscribe());
            });
        });
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgSyncSpacer, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "21.1.2", type: NgSyncSpacer, isStandalone: true, selector: "ng-scrollbar[externalViewport][syncSpacer]", host: { properties: { "style.--_spacer-width": "spacerDimension().width", "style.--_spacer-height": "spacerDimension().height" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgSyncSpacer, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-scrollbar[externalViewport][syncSpacer]',
                    host: {
                        '[style.--_spacer-width]': 'spacerDimension().width',
                        '[style.--_spacer-height]': 'spacerDimension().height'
                    }
                }]
        }], ctorParameters: () => [] });

class NgScrollbarModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarModule, imports: [NgScrollbar,
            NgScrollbarExt,
            NgScrollbarAsyncViewport,
            NgSyncSpacer], exports: [NgScrollbar,
            NgScrollbarExt,
            NgScrollbarAsyncViewport,
            NgSyncSpacer] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        NgScrollbar,
                        NgScrollbarExt,
                        NgScrollbarAsyncViewport,
                        NgSyncSpacer
                    ],
                    exports: [
                        NgScrollbar,
                        NgScrollbarExt,
                        NgScrollbarAsyncViewport,
                        NgSyncSpacer
                    ]
                }]
        }] });
function provideScrollbarOptions(options) {
    return [
        {
            provide: NG_SCROLLBAR_OPTIONS,
            useValue: { ...defaultOptions, ...options }
        }
    ];
}
function provideScrollbarPolyfill(url) {
    return makeEnvironmentProviders([
        {
            provide: NG_SCROLLBAR_POLYFILL,
            useValue: url
        }
    ]);
}

class NgScrollbarAnywhere {
    document = inject(DOCUMENT);
    appRef = inject(ApplicationRef);
    environmentInjector = inject(EnvironmentInjector);
    /**
     * Generic method to create and attach a scrollbar component.
     * @param host - CSS selector for the host element.
     * @param component - The scrollbar component to create (either NgScrollbar or NgScrollbarExt).
     * @param options - Scrollbar options
     * @returns A reference to the created scrollbar component or null if the host is not found.
     */
    createScrollbarComponent(host, component, options, params) {
        const hostElement = typeof host === 'string' ? this.document.querySelector(host) : coerceElement(host);
        if (!hostElement) {
            console.error(`[NgScrollbar]: Could not find the host element!`);
            return null;
        }
        // Create the scrollbar component
        let componentRef = createComponent(component, {
            hostElement,
            environmentInjector: this.environmentInjector,
            projectableNodes: [Array.from(hostElement.childNodes)],
            elementInjector: Injector.create({
                providers: [provideScrollbarOptions(options)]
            }),
            bindings: params ? [
                inputBinding('externalViewport', () => params.viewport),
                inputBinding('externalContentWrapper', () => params.contentWrapper),
                inputBinding('externalSpacer', () => params.spacer),
            ] : null
        });
        // Attach the component's view to the Angular change detection tree
        this.appRef.attachView(componentRef.hostView);
        return {
            componentRef,
            destroy: () => {
                this.appRef.detachView(componentRef.hostView);
                componentRef.destroy();
                componentRef = null;
            }
        };
    }
    /**
     * Creates a basic scrollbar for a given host element.
     * @param host - CSS selector for the target element.
     * @param options - Scrollbar options
     * @returns A reference to the created scrollbar component.
     */
    createScrollbar(host, options) {
        return this.createScrollbarComponent(host, NgScrollbar, options);
    }
    /**
     * Creates an extended scrollbar with external viewport and optional configurations.
     * @param params - Configuration options including host, viewport, and optional elements.
     * @param options - Scrollbar options
     * @returns A reference to the created extended scrollbar component.
     */
    createScrollbarExt(params, options) {
        return this.createScrollbarComponent(params.host, NgScrollbarExt, options, params);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarAnywhere, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarAnywhere, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarAnywhere, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });

class NgScrollbarDocument {
    document = inject(DOCUMENT);
    injector = inject(Injector);
    appRef = inject(ApplicationRef);
    attachScrollbar(options) {
        const viewport = this.document.documentElement;
        const contentWrapper = this.document.body;
        const scrollbarRef = createComponent(ScrollViewport, {
            hostElement: viewport,
            environmentInjector: this.appRef.injector,
            projectableNodes: [Array.from(viewport.childNodes)],
            elementInjector: Injector.create({
                parent: this.injector,
                providers: [
                    ViewportAdapter,
                    provideScrollbarOptions(options)
                ]
            }),
            bindings: [
                inputBinding('contentWrapperElement', () => contentWrapper)
            ]
        });
        this.appRef.attachView(scrollbarRef.hostView);
        return scrollbarRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarDocument, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarDocument, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.1.2", ngImport: i0, type: NgScrollbarDocument, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NG_SCROLLBAR_OPTIONS, NG_SCROLLBAR_POLYFILL, NgScrollbar, NgScrollbarAnywhere, NgScrollbarAsyncViewport, NgScrollbarDocument, NgScrollbarExt, NgScrollbarModule, NgSyncSpacer, ScrollbarInputOutputs, ScrollbarUpdateReason, ViewportAdapter, provideScrollbarOptions, provideScrollbarPolyfill };
